From 2bf4346a4bbdc5da5c3edb0bee63519043b37a33 Mon Sep 17 00:00:00 2001
From: DemonWav <demonwav@gmail.com>
Date: Sat, 16 Sep 2017 23:19:13 -0400
Subject: [PATCH] Tag this as Paper-MCP


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index e9cec44..e76d003 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -490,7 +490,7 @@ public final class CraftServer implements Server {
     }
 
     public String getName() {
-        return "CraftBukkit";
+        return "Paper-MCP";
     }
 
     public String getVersion() {
-- 
2.14.1

