From cde06f7858a238efdd5ce46b19071bca58f4022d Mon Sep 17 00:00:00 2001
From: DemonWav <demonwav@gmail.com>
Date: Sun, 2 Apr 2017 23:10:15 -0500
Subject: [PATCH] More decomp fixes


diff --git a/src/main/java/net/minecraft/world/WorldServer.java b/src/main/java/net/minecraft/world/WorldServer.java
index 207209c..e0937ad 100644
--- a/src/main/java/net/minecraft/world/WorldServer.java
+++ b/src/main/java/net/minecraft/world/WorldServer.java
@@ -1349,12 +1349,14 @@ public class WorldServer extends World implements IThreadListener {
         return this.getChunkProvider().getStrongholdGen(this, p_190528_1_, p_190528_2_, p_190528_3_);
     }
 
+    /* // Paper - fix decompile start
     /**
      * gets the world's chunk provider
-     */
+     * /
     public IChunkProvider getChunkProvider() {
         return this.getChunkProvider();
     }
+    // Paper - fix decompile end */
 
     static class ServerBlockEventList extends ArrayList<BlockEventData> {
         private ServerBlockEventList() {
@@ -1364,4 +1366,4 @@ public class WorldServer extends World implements IThreadListener {
             this();
         }
     }
-}
\ No newline at end of file
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 3204493..28f62c9 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -125,7 +125,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private double health = 20.0D;
     private boolean scaledHealth = false;
     private double healthScale = 20.0D;
-    private final Spigot spigot = new Spigot() {
+    private final Player.Spigot spigot = new Player.Spigot() { // Paper - fix decompile
         public InetSocketAddress getRawAddress() {
             return (InetSocketAddress)CraftPlayer.this.getHandle().connection.netManager.getRawAddress();
         }
@@ -1402,7 +1402,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
     }
 
-    public Spigot spigot() {
+    public Player.Spigot spigot() { // Paper - fix decompile
         return this.spigot;
     }
-}
\ No newline at end of file
+}
-- 
2.12.1

