From 3de1d1ac681bea0bac2938332361b3b6f7846d37 Mon Sep 17 00:00:00 2001
From: DemonWav <demonwav@gmail.com>
Date: Sun, 2 Apr 2017 23:18:59 -0500
Subject: [PATCH] More decompile


diff --git a/src/main/java/net/minecraft/block/properties/PropertyEnum.java b/src/main/java/net/minecraft/block/properties/PropertyEnum.java
index 5fc0560..e9b423c 100644
--- a/src/main/java/net/minecraft/block/properties/PropertyEnum.java
+++ b/src/main/java/net/minecraft/block/properties/PropertyEnum.java
@@ -26,7 +26,7 @@ public class PropertyEnum<T extends Enum<T> & IStringSerializable> extends Prope
                 throw new IllegalArgumentException("Multiple values have the same name \'" + var6 + "\'");
             }
 
-            this.nameToValue.put(var6, var5);
+            this.nameToValue.put(var6, (T) var5); // Paper - fix decompile
         }
     }
 
diff --git a/src/main/java/net/minecraft/command/CommandHandler.java b/src/main/java/net/minecraft/command/CommandHandler.java
index 92bb29c..d5336cb 100644
--- a/src/main/java/net/minecraft/command/CommandHandler.java
+++ b/src/main/java/net/minecraft/command/CommandHandler.java
@@ -53,7 +53,7 @@ public abstract class CommandHandler implements ICommandManager {
                 sender.sendMessage(var15);
             } else if (var5.checkPermission(this.getServer(), sender)) {
                 if (var7 > -1) {
-                    List var16 = EntitySelector.matchEntities(sender, var3[var7], Entity.class);
+                    List<Entity> var16 = EntitySelector.matchEntities(sender, var3[var7], Entity.class); // Paper - fix decompile
                     String var9 = var3[var7];
                     sender.setCommandStat(CommandResultStats.Type.AFFECTED_ENTITIES, var16.size());
 
diff --git a/src/main/java/net/minecraft/entity/ai/EntityAINearestAttackableTarget.java b/src/main/java/net/minecraft/entity/ai/EntityAINearestAttackableTarget.java
index 7e8de96..710c5d8 100644
--- a/src/main/java/net/minecraft/entity/ai/EntityAINearestAttackableTarget.java
+++ b/src/main/java/net/minecraft/entity/ai/EntityAINearestAttackableTarget.java
@@ -66,7 +66,7 @@ public class EntityAINearestAttackableTarget<T extends EntityLivingBase> extends
                 return false;
             } else {
                 Collections.sort(list, this.theNearestAttackableTargetSorter);
-                this.targetEntity = (EntityLivingBase)list.get(0);
+                this.targetEntity = (T)list.get(0); // Paper - fix decompile
                 return true;
             }
         } else {
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 3659868..3501d83 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -768,9 +768,11 @@ public class DedicatedServer extends MinecraftServer implements IServer {
         }
     }
 
+    /* // Paper - fix decompile
     public PlayerList getPlayerList() {
         return this.getPlayerList();
     }
+    // Paper - fix decompile */
 
     public PropertyManager getPropertyManager() {
         return this.settings;
-- 
2.12.1

