From c5d6cf718f7d838349c06c14f52af9ac7340ce44 Mon Sep 17 00:00:00 2001
From: DemonWav <demonwav@gmail.com>
Date: Sun, 2 Apr 2017 23:48:40 -0500
Subject: [PATCH] Decomp


diff --git a/src/main/java/net/minecraft/world/World.java b/src/main/java/net/minecraft/world/World.java
index 7e2c001..20e5c7f 100644
--- a/src/main/java/net/minecraft/world/World.java
+++ b/src/main/java/net/minecraft/world/World.java
@@ -2576,7 +2576,7 @@ public abstract class World implements IBlockAccess {
         ArrayList arraylist = Lists.newArrayList();
 
         for (Entity entity : this.loadedEntityList) {
-            if (entityType.isAssignableFrom(entity.getClass()) && filter.apply(entity)) {
+            if (entityType.isAssignableFrom(entity.getClass()) && filter.apply((T) entity)) { // Paper - fix decompile
                 arraylist.add(entity);
             }
         }
@@ -2588,7 +2588,7 @@ public abstract class World implements IBlockAccess {
         ArrayList arraylist = Lists.newArrayList();
 
         for (Entity entity : this.playerEntities) {
-            if (playerType.isAssignableFrom(entity.getClass()) && filter.apply(entity)) {
+            if (playerType.isAssignableFrom(entity.getClass()) && filter.apply((T) entity)) {  // Paper - fix decompile
                 arraylist.add(entity);
             }
         }
-- 
2.12.1

