From 7c8f7c6ce77041d8797d86d3c4f21fe59a31e187 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Tue, 1 Mar 2016 12:45:11 -0600
Subject: [PATCH] Configurable squid spawn ranges


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 1d59136..d12ed04 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -62,4 +62,12 @@ public class PaperWorldConfig {
         config.addDefault("world-settings.default." + path, def);
         return config.getString("world-settings." + worldName + "." + path, config.getString("world-settings.default." + path));
     }
+
+    public double squidMinSpawnHeight;
+    public double squidMaxSpawnHeight;
+    private void squidSpawnHeights() {
+        squidMinSpawnHeight = getDouble("squid-spawn-height.minimum", 45.0D);
+        squidMaxSpawnHeight = getDouble("squid-spawn-height.maximum", 63.0D);
+        log("Squids will spawn between Y: " + squidMinSpawnHeight + " and Y: " + squidMaxSpawnHeight);
+    }
 }
diff --git a/src/main/java/net/minecraft/entity/passive/EntitySquid.java b/src/main/java/net/minecraft/entity/passive/EntitySquid.java
index 3754d5e..ec40d77 100644
--- a/src/main/java/net/minecraft/entity/passive/EntitySquid.java
+++ b/src/main/java/net/minecraft/entity/passive/EntitySquid.java
@@ -176,7 +176,7 @@ public class EntitySquid extends EntityWaterMob {
      * Checks if the entity's current position is a valid location to spawn this entity.
      */
     public boolean getCanSpawnHere() {
-        return this.posY > 45.0D && this.posY < (double)this.world.getSeaLevel() && super.getCanSpawnHere();
+        return this.posY > world.paperConfig.squidMinSpawnHeight && this.posY < world.paperConfig.squidMaxSpawnHeight && super.getCanSpawnHere(); // Paper
     }
 
     public void setMovementVector(float randomMotionVecXIn, float randomMotionVecYIn, float randomMotionVecZIn) {
-- 
2.9.3

