From aac7e3727c90b9923de22d749842090bc878b200 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sat, 23 Sep 2017 20:49:37 -0400
Subject: [PATCH] Configurable cactus and reed natural growth heights


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index f248393..6228bc1 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -71,4 +71,13 @@ public class PaperWorldConfig {
         squidMaxSpawnHeight = getDouble("squid-spawn-height.maximum", 63.0D);
         log("Squids will spawn between Y: " + squidMinSpawnHeight + " and Y: " + squidMaxSpawnHeight);
     }
+
+    public int cactusMaxHeight;
+    public int reedMaxHeight;
+    private void blowGrowthHeight() {
+        cactusMaxHeight = getInt("max-growth-height.cactus", 3);
+        reedMaxHeight = getInt("max-growth-height.reeds", 3);
+        log("Max height for cactus growth " + cactusMaxHeight + ". Max height for reed growth " + reedMaxHeight);
+
+    }
 }
diff --git a/src/main/java/net/minecraft/block/BlockCactus.java b/src/main/java/net/minecraft/block/BlockCactus.java
index 28e988c..16438f6 100644
--- a/src/main/java/net/minecraft/block/BlockCactus.java
+++ b/src/main/java/net/minecraft/block/BlockCactus.java
@@ -38,7 +38,7 @@ public class BlockCactus extends Block {
             for (i = 1; worldIn.getBlockState(pos.down(i)).getBlock() == this; ++i) {;
             }
 
-            if (i < 3) {
+            if (i < worldIn.paperConfig.cactusMaxHeight) { // Paper - Configurable growth height
                 int j = ((Integer) state.getValue(AGE)).intValue();
 
                 if (j >= (byte) ((int) range(3.0F, 100.0F / (float) worldIn.spigotConfig.cactusModifier * 15.0F + 0.5F, 15.0F))) {
diff --git a/src/main/java/net/minecraft/block/BlockReed.java b/src/main/java/net/minecraft/block/BlockReed.java
index 5faa542..c9d7915 100644
--- a/src/main/java/net/minecraft/block/BlockReed.java
+++ b/src/main/java/net/minecraft/block/BlockReed.java
@@ -40,7 +40,7 @@ public class BlockReed extends Block {
             for (i = 1; worldIn.getBlockState(pos.down(i)).getBlock() == this; ++i) {;
             }
 
-            if (i < 3) {
+            if (i < worldIn.paperConfig.reedMaxHeight) { // Paper - Configurable growth height
                 int j = ((Integer) state.getValue(AGE)).intValue();
 
                 if (j >= (byte) ((int) range(3.0F, 100.0F / (float) worldIn.spigotConfig.caneModifier * 15.0F + 0.5F, 15.0F))) {
-- 
2.14.1

