From cf4bc6c9618724276ca13f95205f5c67662a2a02 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sat, 23 Sep 2017 20:51:50 -0400
Subject: [PATCH] Configurable baby zombie movement speed


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 6228bc1..fe9e7b3 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -80,4 +80,10 @@ public class PaperWorldConfig {
         log("Max height for cactus growth " + cactusMaxHeight + ". Max height for reed growth " + reedMaxHeight);
 
     }
+
+    public double babyZombieMovementSpeed;
+    private void babyZombieMovementSpeed() {
+        babyZombieMovementSpeed = getDouble("baby-zombie-movement-speed", 0.5D); // Player moves at 0.1F, for reference
+        log("Baby zombies will move at the speed of " + babyZombieMovementSpeed);
+    }
 }
diff --git a/src/main/java/net/minecraft/entity/monster/EntityZombie.java b/src/main/java/net/minecraft/entity/monster/EntityZombie.java
index 6aafe53..8999e3e 100644
--- a/src/main/java/net/minecraft/entity/monster/EntityZombie.java
+++ b/src/main/java/net/minecraft/entity/monster/EntityZombie.java
@@ -59,7 +59,7 @@ public class EntityZombie extends EntityMob {
             (new RangedAttribute((IAttribute) null, "zombie.spawnReinforcements", 0.0D, 0.0D, 1.0D))
                     .setDescription("Spawn Reinforcements Chance");
     private static final UUID BABY_SPEED_BOOST_ID = UUID.fromString("B9766B59-9566-4402-BC1F-2EE2A276D836");
-    private static final AttributeModifier BABY_SPEED_BOOST = new AttributeModifier(BABY_SPEED_BOOST_ID, "Baby speed boost", 0.5D, 1);
+    private final AttributeModifier BABY_SPEED_BOOST = new AttributeModifier(BABY_SPEED_BOOST_ID, "Baby speed boost", world.paperConfig.babyZombieMovementSpeed, 1); // Paper - Remove static - Make baby speed configurable
     private static final DataParameter<Boolean> IS_CHILD =
             EntityDataManager.<Boolean>createKey(EntityZombie.class, DataSerializers.BOOLEAN);
     private static final DataParameter<Integer> VILLAGER_TYPE =
-- 
2.14.1

