From a908fd680b05d68a67257a4eeee16802187e8900 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Tue, 1 Mar 2016 13:24:16 -0600
Subject: [PATCH] Allow nerfed mobs to jump


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 17337dc..95b2d3b 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -93,4 +93,9 @@ public class PaperWorldConfig {
         fishingMaxTicks = getInt("fishing-time-range.MaximumTicks", 600);
         log("Fishing time ranges are between " + fishingMinTicks + " and " + fishingMaxTicks + " ticks");
     }
+
+    public boolean nerfedMobsShouldJump;
+    private void nerfedMobsShouldJump() {
+        nerfedMobsShouldJump = getBoolean("spawner-nerfed-mobs-should-jump", false);
+    }
 }
diff --git a/src/main/java/net/minecraft/entity/EntityLiving.java b/src/main/java/net/minecraft/entity/EntityLiving.java
index 8851233..638378f 100644
--- a/src/main/java/net/minecraft/entity/EntityLiving.java
+++ b/src/main/java/net/minecraft/entity/EntityLiving.java
@@ -100,6 +100,7 @@ public abstract class EntityLiving extends EntityLivingBase {
     private boolean isLeashed;
     private Entity leashedToEntity;
     private NBTTagCompound leashNBTTag;
+    public net.minecraft.entity.ai.EntityAISwimming goalFloat; // Paper
 
     public EntityLiving(World worldIn) {
         super(worldIn);
@@ -765,6 +766,13 @@ public abstract class EntityLiving extends EntityLivingBase {
             this.jumpHelper.doJump();
             this.world.profiler.endSection();
             this.world.profiler.endSection();
+        } else {
+            // Paper start - Allow nerfed mobs to jump and float
+            if (goalFloat != null) {
+                if (goalFloat.shouldExecute()) goalFloat.updateTask();
+                this.getJumpHelper().doJump();
+            }
+            //Paper end
         }
     }
 
diff --git a/src/main/java/net/minecraft/entity/ai/EntityAISwimming.java b/src/main/java/net/minecraft/entity/ai/EntityAISwimming.java
index bf7efb9..b377ba5 100644
--- a/src/main/java/net/minecraft/entity/ai/EntityAISwimming.java
+++ b/src/main/java/net/minecraft/entity/ai/EntityAISwimming.java
@@ -7,6 +7,7 @@ public class EntityAISwimming extends EntityAIBase {
     private final EntityLiving theEntity;
 
     public EntityAISwimming(EntityLiving entitylivingIn) {
+        if (entitylivingIn.fromMobSpawner && entitylivingIn.getEntityWorld().paperConfig.nerfedMobsShouldJump) entitylivingIn.goalFloat = this; // Paper
         this.theEntity = entitylivingIn;
         this.setMutexBits(4);
         ((PathNavigateGround)entitylivingIn.getNavigator()).setCanSwim(true);
-- 
2.9.3

