From 9de690dba16d725872207ae085071f61e27fc243 Mon Sep 17 00:00:00 2001
From: Suddenly <suddenly@suddenly.coffee>
Date: Tue, 1 Mar 2016 13:51:54 -0600
Subject: [PATCH] Add configurable despawn distances for living entities


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 5f13fbf..4b64cca 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -99,4 +99,20 @@ public class PaperWorldConfig {
     private void nerfedMobsShouldJump() {
         nerfedMobsShouldJump = getBoolean("spawner-nerfed-mobs-should-jump", false);
     }
+
+    public int softDespawnDistance;
+    public int hardDespawnDistance;
+    private void despawnDistances() {
+        softDespawnDistance = getInt("despawn-ranges.soft", 32); // 32^2 = 1024, Minecraft Default
+        hardDespawnDistance = getInt("despawn-ranges.hard", 128); // 128^2 = 16384, Minecraft Default
+
+        if (softDespawnDistance > hardDespawnDistance) {
+            softDespawnDistance = hardDespawnDistance;
+        }
+
+        log("Living Entity Despawn Ranges:  Soft: " + softDespawnDistance + " Hard: " + hardDespawnDistance);
+
+        softDespawnDistance = softDespawnDistance*softDespawnDistance;
+        hardDespawnDistance = hardDespawnDistance*hardDespawnDistance;
+    }
 }
diff --git a/src/main/java/net/minecraft/entity/EntityLiving.java b/src/main/java/net/minecraft/entity/EntityLiving.java
index 5325a78..26b87c3 100644
--- a/src/main/java/net/minecraft/entity/EntityLiving.java
+++ b/src/main/java/net/minecraft/entity/EntityLiving.java
@@ -654,13 +654,13 @@ public abstract class EntityLiving extends EntityLivingBase {
                 double d2 = entityhuman.posZ - this.posZ;
                 double d3 = d0 * d0 + d1 * d1 + d2 * d2;
 
-                if (d3 > 16384.0D) {
+                if (d3 > world.paperConfig.hardDespawnDistance) { // Paper - custom despawn distances
                     this.setDead();
                 }
 
-                if (this.idleTime > 600 && this.rand.nextInt(800) == 0 && d3 > 1024.0D) {
+                if (this.idleTime > 600 && this.rand.nextInt(800) == 0 && d3 > world.paperConfig.softDespawnDistance) { // Paper - custom despawn distances
                     this.setDead();
-                } else if (d3 < 1024.0D) {
+                } else if (d3 < world.paperConfig.softDespawnDistance) { // Paper - custom despawn distances
                     this.idleTime = 0;
                 }
             }
-- 
2.14.1

