From 9209a2de445216046ded70d3fdad744e59cb5263 Mon Sep 17 00:00:00 2001
From: Byteflux <byte@byteflux.net>
Date: Sat, 23 Sep 2017 21:18:01 -0400
Subject: [PATCH] Configurable speed for water flowing over lava


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 8cdeb34..62e6830 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -131,4 +131,10 @@ public class PaperWorldConfig {
         if (fallingBlockHeightNerf != 0) log("Falling Block Height Limit set to Y: " + fallingBlockHeightNerf);
         if (entityTNTHeightNerf != 0) log("TNT Entity Height Limit set to Y: " + entityTNTHeightNerf);
     }
+
+    public int waterOverLavaFlowSpeed;
+    private void waterOverLawFlowSpeed() {
+        waterOverLavaFlowSpeed = getInt("water-over-lava-flow-speed", 5);
+        log("Water over lava flow speed: " + waterOverLavaFlowSpeed);
+    }
 }
diff --git a/src/main/java/net/minecraft/block/BlockDynamicLiquid.java b/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
index 4197fa9..4a8d0b7 100644
--- a/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
+++ b/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
@@ -35,7 +35,7 @@ public class BlockDynamicLiquid extends BlockLiquid {
             b0 = 2;
         }
 
-        int j = this.tickRate(worldIn);
+        int j = this.getFlowSpeed(worldIn, pos); // Paper
 
         if (i > 0) {
             int l = -100;
@@ -252,7 +252,21 @@ public class BlockDynamicLiquid extends BlockLiquid {
 
     public void onBlockAdded(World worldIn, BlockPos pos, IBlockState state) {
         if (!this.checkForMixing(worldIn, pos, state)) {
-            worldIn.scheduleUpdate(pos, this, this.tickRate(worldIn));
+            worldIn.scheduleUpdate(pos, this, this.getFlowSpeed(worldIn, pos)); // Paper
         }
     }
+
+    /**
+     * Paper - Get flow speed. Throttle if its water and flowing adjacent to lava
+     */
+    private int getFlowSpeed(World world, BlockPos pos) {
+        if (this.blockMaterial == Material.WATER && (
+                world.getBlockState(pos.north(1)).getBlock().blockMaterial == Material.LAVA ||
+                        world.getBlockState(pos.south(1)).getBlock().blockMaterial == Material.LAVA ||
+                        world.getBlockState(pos.west(1)).getBlock().blockMaterial == Material.LAVA ||
+                        world.getBlockState(pos.east(1)).getBlock().blockMaterial == Material.LAVA)) {
+            return world.paperConfig.waterOverLavaFlowSpeed;
+        }
+        return super.tickRate(world);
+    }
 }
-- 
2.14.1

