From 0e8e4d35252aee8d90bb6290e386b475810c4412 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Fri, 7 Apr 2017 22:55:05 -0500
Subject: [PATCH] Show 'Paper' in client crashes, server lists, and Mojang
 stats


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 9d53ef4..c33b844 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1035,7 +1035,7 @@ public abstract class MinecraftServer implements Runnable, ICommandSender, IThre
     }
 
     public String getServerModName() {
-        return "Spigot";
+        return "Paper"; // Paper - Paper > Spigot
     }
 
     /**
diff --git a/src/main/java/net/minecraft/server/ServerEula.java b/src/main/java/net/minecraft/server/ServerEula.java
index 2981ede..590fbbd 100644
--- a/src/main/java/net/minecraft/server/ServerEula.java
+++ b/src/main/java/net/minecraft/server/ServerEula.java
@@ -50,7 +50,8 @@ public class ServerEula {
             Properties var2 = new Properties();
             var1 = new FileOutputStream(this.eulaFile);
             var2.setProperty("eula", "false");
-            var2.store((OutputStream)var1, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula).");
+            var2.store((OutputStream)var1, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula)." +
+                    "\nand also agreeing that tacos are tasty.");  // Paper - fix lag
         } catch (Exception var6) {
             LOG.warn("Failed to save {}", new Object[] {this.eulaFile, var6});
         } finally {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 99445ee..1985155 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -179,7 +179,7 @@ import org.yaml.snakeyaml.error.MarkedYAMLException;
 
 public final class CraftServer implements Server {
     private static final Player[] EMPTY_PLAYER_ARRAY = new Player[0];
-    private final String serverName = "CraftBukkit";
+    private final String serverName = "Paper"; // Paper
     private final String serverVersion;
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
diff --git a/src/main/java/org/spigotmc/WatchdogThread.java b/src/main/java/org/spigotmc/WatchdogThread.java
index 71bbe3c..b9dd5ac 100644
--- a/src/main/java/org/spigotmc/WatchdogThread.java
+++ b/src/main/java/org/spigotmc/WatchdogThread.java
@@ -17,7 +17,7 @@ public class WatchdogThread extends Thread {
     private volatile boolean stopping;
 
     private WatchdogThread(long timeoutTime, boolean restart) {
-        super("Spigot Watchdog Thread");
+        super("Paper Watchdog Thread");
         this.timeoutTime = timeoutTime;
         this.restart = restart;
     }
@@ -44,9 +44,9 @@ public class WatchdogThread extends Thread {
             if (this.lastTick != 0L && System.currentTimeMillis() > this.lastTick + this.timeoutTime) {
                 Logger log = Bukkit.getServer().getLogger();
                 log.log(Level.SEVERE, "The server has stopped responding!");
-                log.log(Level.SEVERE, "Please report this to http://www.spigotmc.org/");
+                log.log(Level.SEVERE, "Please report this to https://aquifermc.org/" );
                 log.log(Level.SEVERE, "Be sure to include ALL relevant console errors and Minecraft crash reports");
-                log.log(Level.SEVERE, "Spigot version: " + Bukkit.getServer().getVersion());
+                log.log( Level.SEVERE, "Paper version: " + Bukkit.getServer().getVersion() );
 
                 if (World.haveWeSilencedAPhysicsCrash) {
                     log.log(Level.SEVERE, "------------------------------");
@@ -55,7 +55,7 @@ public class WatchdogThread extends Thread {
                 }
 
                 log.log(Level.SEVERE, "------------------------------");
-                log.log(Level.SEVERE, "Server thread dump (Look for plugins here before reporting to Spigot!):");
+                log.log( Level.SEVERE, "Server thread dump (Look for plugins here before reporting to Paper!):" );
                 dumpThread(ManagementFactory.getThreadMXBean().getThreadInfo(MinecraftServer.getServer().primaryThread.getId(), Integer.MAX_VALUE), log);
                 log.log(Level.SEVERE, "------------------------------");
                 log.log(Level.SEVERE, "Entire Thread Dump:");
-- 
2.9.3

