From 92173c0f6fdf46886571b260818ddec2b92f0664 Mon Sep 17 00:00:00 2001
From: DemonWav <demonwav@gmail.com>
Date: Mon, 3 Apr 2017 03:01:44 -0500
Subject: [PATCH] Fix infinite recursion


diff --git a/src/main/java/net/minecraft/entity/ai/EntityAINearestAttackableTarget.java b/src/main/java/net/minecraft/entity/ai/EntityAINearestAttackableTarget.java
index 0ba9e00..b7195bd 100644
--- a/src/main/java/net/minecraft/entity/ai/EntityAINearestAttackableTarget.java
+++ b/src/main/java/net/minecraft/entity/ai/EntityAINearestAttackableTarget.java
@@ -48,7 +48,7 @@ public class EntityAINearestAttackableTarget<T extends EntityLivingBase> extends
                 return t0 == null ? false : (targetSelector != null && !targetSelector.apply(t0) ? false : (!EntitySelectors.NOT_SPECTATING.apply(t0) ? false : EntityAINearestAttackableTarget.this.isSuitableTarget(t0, false)));
             }
             public boolean apply(@Nullable Object object) {
-                return this.apply((EntityLivingBase)object);
+                return this.apply((T)object); // Paper - fix decompile (infinite recursion ftw)
             }
         };
     }
-- 
2.12.1

