From cb649f609f1eedbfeaeba728c55ab2f87b71abc1 Mon Sep 17 00:00:00 2001
From: Byteflux <byte@byteflux.net>
Date: Tue, 1 Mar 2016 15:08:03 -0600
Subject: [PATCH] Remove invalid mob spawner tile entities


diff --git a/src/main/java/net/minecraft/world/chunk/Chunk.java b/src/main/java/net/minecraft/world/chunk/Chunk.java
index 4bc3e49..220cc2e 100644
--- a/src/main/java/net/minecraft/world/chunk/Chunk.java
+++ b/src/main/java/net/minecraft/world/chunk/Chunk.java
@@ -30,6 +30,7 @@ import net.minecraft.init.Blocks;
 import net.minecraft.inventory.IInventory;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.tileentity.TileEntity;
+import net.minecraft.tileentity.TileEntityMobSpawner;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.util.ReportedException;
 import net.minecraft.util.math.AxisAlignedBB;
@@ -751,6 +752,10 @@ public class Chunk {
 
             tileEntityIn.validate();
             this.tileEntities.put(pos, tileEntityIn);
+            // Paper start - Remove invalid mob spawner tile entities
+        } else if (tileEntityIn instanceof TileEntityMobSpawner && org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(getBlockState(pos).getBlock()) != org.bukkit.Material.MOB_SPAWNER) {
+            this.tileEntities.remove(pos);
+            // Paper end
         } else {
             System.out.println(
                     "Attempted to place a tile entity ("
-- 
2.14.1

