From be70303ea424437fd924d32fea84630648bede4e Mon Sep 17 00:00:00 2001
From: Byteflux <byte@byteflux.net>
Date: Tue, 1 Mar 2016 15:08:03 -0600
Subject: [PATCH] Remove invalid mob spawner tile entities


diff --git a/src/main/java/net/minecraft/world/chunk/Chunk.java b/src/main/java/net/minecraft/world/chunk/Chunk.java
index 82389aa..2b8a24a 100644
--- a/src/main/java/net/minecraft/world/chunk/Chunk.java
+++ b/src/main/java/net/minecraft/world/chunk/Chunk.java
@@ -30,6 +30,7 @@ import net.minecraft.init.Blocks;
 import net.minecraft.inventory.IInventory;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.tileentity.TileEntity;
+import net.minecraft.tileentity.TileEntityMobSpawner;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.util.ReportedException;
 import net.minecraft.util.math.AxisAlignedBB;
@@ -775,6 +776,11 @@ public class Chunk {
 
             tileEntityIn.validate();
             this.chunkTileEntityMap.put(pos, tileEntityIn);
+            // Paper start - Remove invalid mob spawner tile entities
+        } else if (tileEntityIn instanceof TileEntityMobSpawner && org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(getBlockState(pos).getBlock()) != org.bukkit.Material.MOB_SPAWNER) {
+            this.chunkTileEntityMap.remove(pos);
+            tileEntityIn.invalidate();
+            // Paper end
         } else {
             System.out.println("Attempted to place a tile entity (" + tileEntityIn + ") at " + tileEntityIn.pos.getX() + "," + tileEntityIn.pos.getY() + "," + tileEntityIn.pos.getZ() + " (" + CraftMagicNumbers.getMaterial(this.getBlockState(pos).getBlock()) + ") where there was no entity tile!");
             System.out.println("Chunk coordinates: " + this.xPosition * 16 + "," + this.zPosition * 16);
-- 
2.12.1.windows.1

