From 4788d479e3cf84fe050cc7bc90aaa41f46ce6cbf Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 24 Sep 2017 15:31:41 -0400
Subject: [PATCH] Prevent tile entity and entity crashes


diff --git a/src/main/java/net/minecraft/tileentity/TileEntity.java b/src/main/java/net/minecraft/tileentity/TileEntity.java
index bf02111..6b87083 100644
--- a/src/main/java/net/minecraft/tileentity/TileEntity.java
+++ b/src/main/java/net/minecraft/tileentity/TileEntity.java
@@ -190,7 +190,12 @@ public abstract class TileEntity {
                 });
 
         if (this.world != null) {
-            CrashReportCategory.addBlockInfo(reportCategory, this.pos, this.getBlockType(), this.getBlockMetadata());
+            // Paper start - Prevent TileEntity and Entity crashes
+            Block block = this.getBlockType();
+            if (block != null) {
+                CrashReportCategory.addBlockInfo(reportCategory, this.pos, this.getBlockType(), this.getBlockMetadata());
+            }
+            // Paper end
             reportCategory.addDetail(
                     "Actual block type",
                     new ICrashReportDetail() {
diff --git a/src/main/java/net/minecraft/world/World.java b/src/main/java/net/minecraft/world/World.java
index 7fb6216..b8d09c3 100644
--- a/src/main/java/net/minecraft/world/World.java
+++ b/src/main/java/net/minecraft/world/World.java
@@ -1571,10 +1571,12 @@ public abstract class World implements IBlockAccess {
                         entity.tickTimer.stopTiming(); // Paper
                     } catch (Throwable var18) {
                         entity.tickTimer.stopTiming();
-                        CrashReport crashreport1 = CrashReport.makeCrashReport(var18, "Ticking entity");
-                        CrashReportCategory crashreportsystemdetails1 = crashreport1.makeCategory("Entity being ticked");
-                        entity.addEntityCrashInfo(crashreportsystemdetails1);
-                        throw new ReportedException(crashreport1);
+                        // Paper start - Prevent tile entity and entity crashes
+                        System.err.println("Entity threw exception at " + entity.world.getWorld().getName() + ":" + entity.posX + "," + entity.posY + "," + entity.posZ);
+                        var18.printStackTrace();
+                        entity.isDead = true;
+                        continue;
+                        // Paper end
                     }
                 }
 
@@ -1639,10 +1641,13 @@ public abstract class World implements IBlockAccess {
                             ((ITickable) tileentity).update();
                             this.profiler.endSection();
                         } catch (Throwable var16) {
-                            CrashReport crashreport1 = CrashReport.makeCrashReport(var16, "Ticking block entity");
-                            CrashReportCategory crashreportsystemdetails1 = crashreport1.makeCategory("Block entity being ticked");
-                            tileentity.addInfoToCrashReport(crashreportsystemdetails1);
-                            throw new ReportedException(crashreport1);
+                            // Paper start - Prevent tile entity and entity crashes
+                            System.err.println("TileEntity threw exception at " + tileentity.world.getWorld().getName() + ":" + tileentity.pos.getX() + "," + tileentity.pos.getY() + "," + tileentity.pos.getZ());
+                            var16.printStackTrace();
+                            tilesThisCycle--;
+                            this.tickableTileEntities.remove(tileTickPosition--);
+                            continue;
+                            // Paper end
                         } finally {
                             tileentity.tickTimer.stopTiming();
                         }
-- 
2.18.0

