From a3ca8cb45bc45693768540a3e5d2c4f28650f8e8 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Tue, 1 Mar 2016 23:58:50 -0600
Subject: [PATCH] Configurable top of nether void damage


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 46bd479..951e128 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -136,4 +136,9 @@ public class PaperWorldConfig {
         waterOverLavaFlowSpeed = getInt("water-over-lava-flow-speed", 5);
         log("Water over lava flow speed: " + waterOverLavaFlowSpeed);
     }
+
+    public boolean netherVoidTopDamage;
+    private void netherVoidTopDamage() {
+        netherVoidTopDamage = getBoolean("nether-ceiling-void-damage", false);
+    }
 }
diff --git a/src/main/java/net/minecraft/entity/Entity.java b/src/main/java/net/minecraft/entity/Entity.java
index ebba0db..20a7a94 100644
--- a/src/main/java/net/minecraft/entity/Entity.java
+++ b/src/main/java/net/minecraft/entity/Entity.java
@@ -532,9 +532,14 @@ public abstract class Entity implements ICommandSender {
             this.fallDistance *= 0.5F;
         }
 
+        // Paper start - Configurable nether ceiling damage, extracted
+        /*
         if (this.posY < -64.0D) {
             this.kill();
         }
+        */
+        this.updateHeightDamage();
+        // Paper end
 
         if (!this.world.isRemote) {
             this.setFlag(0, this.fire > 0);
@@ -544,6 +549,14 @@ public abstract class Entity implements ICommandSender {
         this.world.profiler.endSection();
     }
 
+    // Paper start - Configurable nether ceiling damage
+    protected void updateHeightDamage() {
+        if (this.posY < -64.0D || this.world.paperConfig.netherVoidTopDamage && this.world.getWorld().getEnvironment() == org.bukkit.World.Environment.NETHER && this.posY >= 128.0D) {
+            this.kill();
+        }
+    }
+    // Paper end
+
     /**
      * Decrements the counter for the remaining time until the entity may use a portal again.
      */
diff --git a/src/main/java/net/minecraft/entity/item/EntityMinecart.java b/src/main/java/net/minecraft/entity/item/EntityMinecart.java
index 091d13c..f61a5be 100644
--- a/src/main/java/net/minecraft/entity/item/EntityMinecart.java
+++ b/src/main/java/net/minecraft/entity/item/EntityMinecart.java
@@ -256,9 +256,14 @@ public abstract class EntityMinecart extends Entity implements IWorldNameable {
             this.setDamage(this.getDamage() - 1.0F);
         }
 
+        // Paper - Configurable nether ceiling damage, extracted
+        /*
         if (this.posY < -64.0D) {
             this.kill();
         }
+        */
+        this.updateHeightDamage();
+        // Paper end
 
         if (!this.world.isRemote && this.world instanceof WorldServer) {
             this.world.profiler.startSection("portal");
-- 
2.12.1.windows.1

