From 2f204b37d3dff37093dc4b786b8587a1810e4654 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Wed, 2 Mar 2016 00:03:55 -0600
Subject: [PATCH] Check online mode before converting and renaming player data


diff --git a/src/main/java/net/minecraft/world/storage/SaveHandler.java b/src/main/java/net/minecraft/world/storage/SaveHandler.java
index 53c0eab..8b5670c 100644
--- a/src/main/java/net/minecraft/world/storage/SaveHandler.java
+++ b/src/main/java/net/minecraft/world/storage/SaveHandler.java
@@ -202,7 +202,7 @@ public class SaveHandler implements ISaveHandler, IPlayerFileData {
             File file = new File(this.playersDirectory, player.getCachedUniqueIdString() + ".dat");
             boolean usingWrongFile = false;
 
-            if (!file.exists()) {
+            if (org.bukkit.Bukkit.getOnlineMode() && !file.exists()) { // Paper - Check online mode first
                 file = new File(this.playersDirectory, UUID.nameUUIDFromBytes(("OfflinePlayer:" + player.getName()).getBytes("UTF-8")).toString() + ".dat");
 
                 if (file.exists()) {
-- 
2.12.1.windows.1

