From a07b4bb1edae2e0c0077a5e11ca5884b67f349e3 Mon Sep 17 00:00:00 2001
From: Iceee <andrew@opticgaming.tv>
Date: Sun, 24 Sep 2017 17:23:50 -0400
Subject: [PATCH] Stop updating flowing block if material has changed


diff --git a/src/main/java/net/minecraft/block/BlockDynamicLiquid.java b/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
index 4a8d0b7..3c34353 100644
--- a/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
+++ b/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
@@ -93,6 +93,7 @@ public class BlockDynamicLiquid extends BlockLiquid {
             this.placeStaticBlock(worldIn, pos, state);
         }
 
+        if (worldIn.getBlockState(pos).getMaterial() != blockMaterial) return; // Paper - Stop updating flowing block if material has changed
         org.bukkit.block.Block source = worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
         IBlockState iblockdata2 = worldIn.getBlockState(pos.down());
 
-- 
2.18.0

