From 04fb9056c31c46e56fc0aafac4bd857341c06bc6 Mon Sep 17 00:00:00 2001
From: Iceee <andrew@opticgaming.tv>
Date: Wed, 2 Mar 2016 12:03:23 -0600
Subject: [PATCH] Stop updating flowing block if material has changed


diff --git a/src/main/java/net/minecraft/block/BlockDynamicLiquid.java b/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
index 742a995..2afdfb4 100644
--- a/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
+++ b/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
@@ -92,6 +92,7 @@ public class BlockDynamicLiquid extends BlockLiquid {
             this.placeStaticBlock(worldIn, pos, state);
         }
 
+        if (worldIn.getBlockState(pos).getMaterial() != blockMaterial) return; // Paper - Stop updating flowing block if material has changed
         org.bukkit.block.Block source = worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
         IBlockState iblockdata2 = worldIn.getBlockState(pos.down());
 
-- 
2.12.1.windows.1

