From 36f3a39ee609def5398a2cc27b4b2d2cb3a42607 Mon Sep 17 00:00:00 2001
From: Byteflux <byte@byteflux.net>
Date: Sun, 24 Sep 2017 17:52:09 -0400
Subject: [PATCH] Configurable lava flow speed


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 8a28def..53597a5 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -186,4 +186,11 @@ public class PaperWorldConfig {
         fastDrainLava = getBoolean("fast-drain.lava", false);
         fastDrainWater = getBoolean("fast-drain.water", false);
     }
+
+    public int lavaFlowSpeedNormal;
+    public int lavaFlowSpeedNether;
+    private void lavaFlowSpeeds() {
+        lavaFlowSpeedNormal = getInt("lava-flow-speed.normal", 30);
+        lavaFlowSpeedNether = getInt("lava-flow-speed.nether", 10);
+    }
 }
diff --git a/src/main/java/net/minecraft/block/BlockDynamicLiquid.java b/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
index 2215ab2..e42db27 100644
--- a/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
+++ b/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
@@ -260,6 +260,9 @@ public class BlockDynamicLiquid extends BlockLiquid {
      * Paper - Get flow speed. Throttle if its water and flowing adjacent to lava
      */
     private int getFlowSpeed(World world, BlockPos pos) {
+        if (this.blockMaterial == Material.LAVA) {
+            return world.provider.isNether() ? world.paperConfig.lavaFlowSpeedNether : world.paperConfig.lavaFlowSpeedNormal;
+        }
         if (this.blockMaterial == Material.WATER && (
                 world.getBlockState(pos.north(1)).getBlock().blockMaterial == Material.LAVA ||
                         world.getBlockState(pos.south(1)).getBlock().blockMaterial == Material.LAVA ||
-- 
2.18.0

