From f83375f1202240f072423e7750ba323d8076a51f Mon Sep 17 00:00:00 2001
From: Byteflux <byte@byteflux.net>
Date: Wed, 2 Mar 2016 12:27:07 -0600
Subject: [PATCH] Configurable lava flow speed


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 919d8eb..3e66b40 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -191,4 +191,11 @@ public class PaperWorldConfig {
         fastDrainLava = getBoolean("fast-drain.lava", false);
         fastDrainWater = getBoolean("fast-drain.water", false);
     }
+
+    public int lavaFlowSpeedNormal;
+    public int lavaFlowSpeedNether;
+    private void lavaFlowSpeeds() {
+        lavaFlowSpeedNormal = getInt("lava-flow-speed.normal", 30);
+        lavaFlowSpeedNether = getInt("lava-flow-speed.nether", 10);
+    }
 }
diff --git a/src/main/java/net/minecraft/block/BlockDynamicLiquid.java b/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
index 1f3c14c..3eb4d7e 100644
--- a/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
+++ b/src/main/java/net/minecraft/block/BlockDynamicLiquid.java
@@ -258,6 +258,9 @@ public class BlockDynamicLiquid extends BlockLiquid {
      * Paper - Gets how many ticks before this block is ticked
      */
     public int getTickRate(World world, BlockPos pos) {
+        if (this.blockMaterial == Material.LAVA) {
+            return world.provider.hasNoSky() ? world.paperConfig.lavaFlowSpeedNether : world.paperConfig.lavaFlowSpeedNormal;
+        }
         if (this.blockMaterial == Material.WATER && (
                 world.getBlockState(pos.north(1)).getBlock().blockMaterial == Material.LAVA ||
                         world.getBlockState(pos.south(1)).getBlock().blockMaterial == Material.LAVA ||
-- 
2.12.1.windows.1

