From df38f113c8982fd0bb98360c1dc85d67a49713ee Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sat, 8 Apr 2017 23:03:19 -0500
Subject: [PATCH] Disable explosion knockback


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 6550e46..fed7fde 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -198,4 +198,9 @@ public class PaperWorldConfig {
         lavaFlowSpeedNormal = getInt("lava-flow-speed.normal", 30);
         lavaFlowSpeedNether = getInt("lava-flow-speed.nether", 10);
     }
+    
+    public boolean disableExplosionKnockback;
+    private void disableExplosionKnockback() {
+        disableExplosionKnockback = getBoolean("disable-explosion-knockback", false);
+    }
 }
diff --git a/src/main/java/net/minecraft/entity/EntityLivingBase.java b/src/main/java/net/minecraft/entity/EntityLivingBase.java
index c335e28..9cde0cc 100644
--- a/src/main/java/net/minecraft/entity/EntityLivingBase.java
+++ b/src/main/java/net/minecraft/entity/EntityLivingBase.java
@@ -960,12 +960,14 @@ public abstract class EntityLivingBase extends Entity {
                     }
                 }
 
+                boolean knockbackCancelled = world.paperConfig.disableExplosionKnockback && source.isExplosion() && this instanceof EntityPlayer; // Paper - Disable explosion knockback
                 if (flag1) {
                     if (flag) {
                         this.world.setEntityState(this, (byte)29);
                     } else if (source instanceof EntityDamageSource && ((EntityDamageSource)source).getIsThornsDamage()) {
                         this.world.setEntityState(this, (byte)33);
                     } else {
+                        if (!knockbackCancelled) // Paper start - Disable explosion knockback
                         this.world.setEntityState(this, (byte)2);
                     }
 
@@ -988,6 +990,8 @@ public abstract class EntityLivingBase extends Entity {
                     }
                 }
 
+                if (knockbackCancelled) this.world.setEntityState(this, (byte) 2); // Paper - Disable explosion knockback
+
                 if (this.getHealth() <= 0.0F) {
                     if (!this.checkTotemDeathProtection(source)) {
                         SoundEvent soundeffect = this.getDeathSound();
diff --git a/src/main/java/net/minecraft/world/Explosion.java b/src/main/java/net/minecraft/world/Explosion.java
index daa0ac1..df76233 100644
--- a/src/main/java/net/minecraft/world/Explosion.java
+++ b/src/main/java/net/minecraft/world/Explosion.java
@@ -144,7 +144,7 @@ public class Explosion {
                                 double d14 = d13;
 
                                 if (entity instanceof EntityLivingBase) {
-                                    d14 = EnchantmentProtection.getBlastDamageReduction((EntityLivingBase)entity, d13);
+                                    d14 = entity instanceof EntityPlayer && world.paperConfig.disableExplosionKnockback ? 0 : EnchantmentProtection.getBlastDamageReduction((EntityLivingBase)entity, d13);  // Paper - Disable explosion knockback
                                 }
 
                                 entity.motionX += d8 * d14;
@@ -154,7 +154,7 @@ public class Explosion {
                                 if (entity instanceof EntityPlayer) {
                                     EntityPlayer entityhuman = (EntityPlayer)entity;
 
-                                    if (!entityhuman.isSpectator() && (!entityhuman.isCreative() || !entityhuman.capabilities.isFlying)) {
+                                    if (!entityhuman.isSpectator() && (!entityhuman.isCreative() && !world.paperConfig.disableExplosionKnockback || !entityhuman.capabilities.isFlying)) { // Paper - Disable explosion knockback
                                         this.playerKnockbackMap.put(entityhuman, new Vec3d(d8 * d13, d9 * d13, d10 * d13));
                                     }
                                 }
-- 
2.12.1.windows.1

