From 7859783564fc4afbb3235774ca7cb4963c3e769b Mon Sep 17 00:00:00 2001
From: Sudzzy <originmc@outlook.com>
Date: Sun, 24 Sep 2017 18:32:32 -0400
Subject: [PATCH] Option to disable snow and ice formation


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 9f632a4..f52952e 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -203,4 +203,9 @@ public class PaperWorldConfig {
     private void disableThunder() {
         disableThunder = getBoolean("disable-thunder", false);
     }
+
+    public boolean disableIceAndSnow;
+    private void disableIceAndSnow(){
+        disableIceAndSnow = getBoolean("disable-ice-and-snow", false);
+    }
 }
diff --git a/src/main/java/net/minecraft/world/WorldServer.java b/src/main/java/net/minecraft/world/WorldServer.java
index 2fd47f8..16a044f 100644
--- a/src/main/java/net/minecraft/world/WorldServer.java
+++ b/src/main/java/net/minecraft/world/WorldServer.java
@@ -587,7 +587,7 @@ public class WorldServer extends World implements IThreadListener {
 
                     this.profiler.endStartSection("iceandsnow");
 
-                    if (this.rand.nextInt(16) == 0) {
+                    if (!this.paperConfig.disableIceAndSnow && this.rand.nextInt(16) == 0) { // Paper - Option to disable ice and snow
                         this.updateLCG = this.updateLCG * 3 + 1013904223;
                         int l = this.updateLCG >> 2;
                         BlockPos blockposition = this.getPrecipitationHeight(new BlockPos(j + (l & 15), 0, k + (l >> 8 & 15)));
-- 
2.18.0

