From f4a57c2a4dc29e2d7dc9e6ad13c61040ecb39657 Mon Sep 17 00:00:00 2001
From: Sudzzy <originmc@outlook.com>
Date: Wed, 27 Sep 2017 14:07:39 -0400
Subject: [PATCH] Configurable mob spawner tick rate


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 07f54ff..a1fffc8 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -216,4 +216,9 @@ public class PaperWorldConfig {
     private void disableIceAndSnow(){
         disableIceAndSnow = getBoolean("disable-ice-and-snow", false);
     }
+
+    public int mobSpawnerTickRate;
+    private void mobSpawnerTickRate() {
+        mobSpawnerTickRate = getInt("mob-spawner-tick-rate", 1);
+    }
 }
diff --git a/src/main/java/net/minecraft/tileentity/MobSpawnerBaseLogic.java b/src/main/java/net/minecraft/tileentity/MobSpawnerBaseLogic.java
index c5a1256..2e50695 100644
--- a/src/main/java/net/minecraft/tileentity/MobSpawnerBaseLogic.java
+++ b/src/main/java/net/minecraft/tileentity/MobSpawnerBaseLogic.java
@@ -33,6 +33,7 @@ public abstract class MobSpawnerBaseLogic {
     private int maxNearbyEntities = 6;
     private int activatingRangeFromPlayer = 16;
     private int spawnRange = 4;
+    private int tickDelay = 0; // Paper
 
     @Nullable
     public ResourceLocation getEntityId() {
@@ -58,6 +59,10 @@ public abstract class MobSpawnerBaseLogic {
     }
 
     public void updateSpawner() {
+        // Paper start - Configurable mob spawner tick rate
+        if (spawnDelay > 0 && --tickDelay > 0) return;
+        tickDelay = this.getSpawnerWorld().paperConfig.mobSpawnerTickRate;
+        // Paper end
         if (!this.isActivated()) {
             this.prevMobRotation = this.mobRotation;
         } else {
@@ -71,18 +76,18 @@ public abstract class MobSpawnerBaseLogic {
                 this.getSpawnerWorld().spawnParticle(EnumParticleTypes.FLAME, d0, d1, d2, 0.0D, 0.0D, 0.0D);
 
                 if (this.spawnDelay > 0) {
-                    --this.spawnDelay;
+                    this.spawnDelay -= tickDelay; // Paper
                 }
 
                 this.prevMobRotation = this.mobRotation;
                 this.mobRotation = (this.mobRotation + (double) (1000.0F / ((float) this.spawnDelay + 200.0F))) % 360.0D;
             } else {
-                if (this.spawnDelay == -1) {
+                if (this.spawnDelay < -tickDelay) { // Paper
                     this.resetTimer();
                 }
 
                 if (this.spawnDelay > 0) {
-                    --this.spawnDelay;
+                    this.spawnDelay -= tickDelay; // Paper
                     return;
                 }
 
-- 
2.14.1.windows.1

