From 768e6e07654ea1f9d0c79002cab20d18803acc5d Mon Sep 17 00:00:00 2001
From: Sudzzy <originmc@outlook.com>
Date: Wed, 27 Sep 2017 14:39:45 -0400
Subject: [PATCH] Configurable container update tick rate


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index a1fffc8..b8bf46c 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -221,4 +221,9 @@ public class PaperWorldConfig {
     private void mobSpawnerTickRate() {
         mobSpawnerTickRate = getInt("mob-spawner-tick-rate", 1);
     }
+
+    public int containerUpdateTickRate;
+    private void containerUpdateTickRate() {
+        containerUpdateTickRate = getInt("container-update-tick-rate", 1);
+    }
 }
diff --git a/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java b/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java
index 750f86c..de4563b 100644
--- a/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java
+++ b/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java
@@ -199,6 +199,7 @@ public class EntityPlayerMP extends EntityPlayer implements IContainerListener {
         this.viewDistance = viewDistance;
     }
     // Paper end
+    private int containerUpdateDelay; // Paper
 
     public EntityPlayerMP(MinecraftServer server, WorldServer worldIn, GameProfile profile, PlayerInteractionManager interactionManagerIn) {
         super(worldIn, profile);
@@ -387,7 +388,12 @@ public class EntityPlayerMP extends EntityPlayer implements IContainerListener {
             --this.hurtResistantTime;
         }
 
-        this.openContainer.detectAndSendChanges();
+        // Paper start - Configurable container update tick rate
+        if (--containerUpdateDelay <= 0) {
+            this.openContainer.detectAndSendChanges();
+            containerUpdateDelay = this.world.paperConfig.containerUpdateTickRate;
+        }
+        // Paper end
 
         if (!this.world.isRemote && !this.openContainer.canInteractWith(this)) {
             this.closeScreen();
-- 
2.14.1.windows.1

