From e80f54c1bfcdc5085eb81d6ee1ff10a4e7a81f2d Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@outlook.com>
Date: Wed, 27 Sep 2017 14:39:52 -0400
Subject: [PATCH] Use UserCache for player head


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
index 9712750..a57f3de 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
@@ -6,9 +6,12 @@ import com.mojang.authlib.GameProfile;
 import java.util.Locale;
 import java.util.Map;
 import java.util.UUID;
+
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.nbt.NBTBase;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTUtil;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.tileentity.TileEntitySkull;
 import org.bukkit.Bukkit;
 import org.bukkit.Material;
@@ -134,7 +137,11 @@ class CraftMetaSkull extends CraftMetaItem implements SkullMeta {
             if (name == null) {
                 this.profile = null;
             } else {
-                this.profile = (GameProfile) TileEntitySkull.skinCache.getIfPresent(name.toLowerCase(Locale.ROOT));
+                // Paper start - Use Online Players Skull
+                EntityPlayerMP player = MinecraftServer.getServer().getPlayerList().getPlayerByUsername(name);
+                if (profile == null && player != null) profile = player.getGameProfile();
+                // Paper end
+                if (profile == null) this.profile = (GameProfile) TileEntitySkull.skinCache.getIfPresent(name.toLowerCase(Locale.ROOT));
 
                 if (this.profile == null) {
                     this.profile = new GameProfile((UUID) null, name);
-- 
2.14.1.windows.1

