From 5697e76ff9f01b7979571730c7b13542d1df563d Mon Sep 17 00:00:00 2001
From: Sudzzy <originmc@outlook.com>
Date: Wed, 2 Mar 2016 23:34:44 -0600
Subject: [PATCH] Configurable container update tick rate


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 92a7d61..3fcc4ee 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -218,4 +218,9 @@ public class PaperWorldConfig {
     private void mobSpawnerTickRate() {
         mobSpawnerTickRate = getInt("mob-spawner-tick-rate", 1);
     }
+
+    public int containerUpdateTickRate;
+    private void containerUpdateTickRate() {
+        containerUpdateTickRate = getInt("container-update-tick-rate", 1);
+    }
 }
diff --git a/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java b/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java
index 4475ab2..d80f00b 100644
--- a/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java
+++ b/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java
@@ -206,6 +206,7 @@ public class EntityPlayerMP extends EntityPlayer implements IContainerListener {
         this.viewDistance = viewDistance;
     }
     // Paper end
+    private int containerUpdateDelay; // Paper
 
     public EntityPlayerMP(MinecraftServer server, WorldServer worldIn, GameProfile profile, PlayerInteractionManager interactionManagerIn) {
         super(worldIn, profile);
@@ -379,7 +380,12 @@ public class EntityPlayerMP extends EntityPlayer implements IContainerListener {
             --this.hurtResistantTime;
         }
 
-        this.openContainer.detectAndSendChanges();
+        // Paper start - Configurable container update tick rate
+        if (--containerUpdateDelay <= 0) {
+            this.openContainer.detectAndSendChanges();
+            containerUpdateDelay = world.paperConfig.containerUpdateTickRate;
+        }
+        // Paper end
 
         if (!this.world.isRemote && !this.openContainer.canInteractWith(this)) {
             this.closeScreen();
-- 
2.12.1.windows.1

