From f00931655effda4246a4873d3b5fa3f1ace57093 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Wed, 27 Sep 2017 14:43:33 -0400
Subject: [PATCH] Disable spigot tick limiters


diff --git a/src/main/java/net/minecraft/world/World.java b/src/main/java/net/minecraft/world/World.java
index 602472c..c83282b 100644
--- a/src/main/java/net/minecraft/world/World.java
+++ b/src/main/java/net/minecraft/world/World.java
@@ -1539,9 +1539,9 @@ public abstract class World implements IBlockAccess {
             this.tickPosition = 0;
         }
 
-        this.entityLimiter.initTick();
+        //this.entityLimiter.initTick();
 
-        while (entitiesThisCycle < this.loadedEntityList.size() && (entitiesThisCycle % 10 != 0 || this.entityLimiter.shouldContinue())) {
+        while (entitiesThisCycle < this.loadedEntityList.size()) { // Paper - Disable tick limiters
             label374:
             {
                 this.tickPosition = this.tickPosition < this.loadedEntityList.size() ? this.tickPosition : 0;
@@ -1611,9 +1611,9 @@ public abstract class World implements IBlockAccess {
 
         this.processingLoadedTiles = true;
         int tilesThisCycle = 0;
-        this.tileLimiter.initTick();
+        //this.tileLimiter.initTick();
 
-        while (tilesThisCycle < this.tickableTileEntities.size() && (tilesThisCycle % 10 != 0 || this.tileLimiter.shouldContinue())) {
+        while (tilesThisCycle < this.tickableTileEntities.size()) { // Paper - Disable tick limiters
             this.tileTickPosition = this.tileTickPosition < this.tickableTileEntities.size() ? this.tileTickPosition : 0;
             TileEntity tileentity = this.tickableTileEntities.get(this.tileTickPosition);
 
-- 
2.14.1.windows.1

