From 8b918b4eb67a2e9738bd15635c9f7c73067db721 Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@outlook.com>
Date: Wed, 27 Sep 2017 14:39:52 -0400
Subject: [PATCH] Use UserCache for player head


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
index 29bd40b..dc0eb6e 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
@@ -4,11 +4,11 @@ import com.google.common.base.Predicates;
 import com.google.common.collect.ImmutableMap.Builder;
 import com.google.common.util.concurrent.Futures;
 import com.mojang.authlib.GameProfile;
-import java.util.Map;
-import java.util.UUID;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.nbt.NBTBase;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTUtil;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.tileentity.TileEntitySkull;
 import org.bukkit.Bukkit;
 import org.bukkit.Material;
@@ -16,6 +16,9 @@ import org.bukkit.OfflinePlayer;
 import org.bukkit.configuration.serialization.DelegateDeserialization;
 import org.bukkit.inventory.meta.SkullMeta;
 
+import java.util.Map;
+import java.util.UUID;
+
 @DelegateDeserialization(CraftMetaItem.SerializableMeta.class)
 class CraftMetaSkull extends CraftMetaItem implements SkullMeta {
     static final CraftMetaItem.ItemMetaKey SKULL_PROFILE = new CraftMetaItem.ItemMetaKey("SkullProfile");
@@ -125,7 +128,11 @@ class CraftMetaSkull extends CraftMetaItem implements SkullMeta {
             if (name == null) {
                 this.profile = null;
             } else {
-                this.profile = new GameProfile((UUID) null, name);
+                // Paper start - Use Online Players Skull
+                EntityPlayerMP player = MinecraftServer.getServer().getPlayerList().getPlayerByUsername(name);
+                if (profile == null && player != null) profile = player.getGameProfile();
+                if (profile == null) profile = new GameProfile(null, name);
+                // Paper end
             }
 
             return true;
-- 
2.18.0

