From 3ae7d86e8f7e2a12443ca4996149b2aa1b763761 Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@outlook.com>
Date: Thu, 3 Mar 2016 02:32:10 -0600
Subject: [PATCH] Use UserCache for player heads


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
index 8d54ad2..66036b0 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
@@ -6,9 +6,12 @@ import com.mojang.authlib.GameProfile;
 import java.util.Locale;
 import java.util.Map;
 import java.util.UUID;
+
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.nbt.NBTBase;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTUtil;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.tileentity.TileEntitySkull;
 import org.bukkit.Material;
 import org.bukkit.configuration.serialization.DelegateDeserialization;
@@ -116,7 +119,11 @@ class CraftMetaSkull extends CraftMetaItem implements SkullMeta {
             if (name == null) {
                 this.profile = null;
             } else {
-                this.profile = (GameProfile)TileEntitySkull.skinCache.getIfPresent(name.toLowerCase(Locale.ROOT));
+                // Paper start - Use Online Players Skull
+                EntityPlayerMP player = MinecraftServer.getServer().getPlayerList().getPlayerByUsername(name);
+                if (profile == null && player != null) profile = player.getGameProfile();
+                if (profile == null) this.profile = (GameProfile)TileEntitySkull.skinCache.getIfPresent(name.toLowerCase(Locale.ROOT));
+                // Paper end
 
                 if (this.profile == null) {
                     this.profile = new GameProfile((UUID)null, name);
-- 
2.12.1.windows.1

