From c59d4ab88f27ac30ab752692bf01f1234fc61a4c Mon Sep 17 00:00:00 2001
From: vemacs <d@nkmem.es>
Date: Thu, 3 Mar 2016 01:19:22 -0600
Subject: [PATCH] All chunks are slime spawn chunks toggle


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index ff9919e..ff8230a 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -228,4 +228,9 @@ public class PaperWorldConfig {
     private void disableChestCatDetection() {
         disableChestCatDetection = getBoolean("game-mechanics.disable-chest-cat-detection", false);
     }
+    
+    public boolean allChunksAreSlimeChunks;
+    public void allChunksAreSlimeChunks() {
+        allChunksAreSlimeChunks = getBoolean("all-chunks-are-slime-chunks", false);
+    }
 }
diff --git a/src/main/java/net/minecraft/entity/monster/EntitySlime.java b/src/main/java/net/minecraft/entity/monster/EntitySlime.java
index 4fd7b11..c0450ed 100644
--- a/src/main/java/net/minecraft/entity/monster/EntitySlime.java
+++ b/src/main/java/net/minecraft/entity/monster/EntitySlime.java
@@ -314,7 +314,10 @@ public class EntitySlime extends EntityLiving implements IMob {
                     return super.getCanSpawnHere();
                 }
 
-                if (this.rand.nextInt(10) == 0 && chunk.getRandomWithSeed((long)this.world.spigotConfig.slimeSeed).nextInt(10) == 0 && this.posY < 40.0D) {
+                // Paper start
+                boolean isSlimeChunk = world.paperConfig.allChunksAreSlimeChunks || chunk.getRandomWithSeed((long)this.world.spigotConfig.slimeSeed).nextInt(10) == 0;
+                if (this.rand.nextInt(10) == 0 && isSlimeChunk && this.posY < 40.0D) {
+                    // Paper end
                     return super.getCanSpawnHere();
                 }
             }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index bffb129..5ccbf0b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -125,7 +125,7 @@ public class CraftChunk implements Chunk {
     }
 
     public boolean isSlimeChunk() {
-        return this.getHandle().getRandomWithSeed((long)this.worldServer.spigotConfig.slimeSeed).nextInt(10) == 0;
+        return worldServer.paperConfig.allChunksAreSlimeChunks || this.getHandle().getRandomWithSeed((long)this.worldServer.spigotConfig.slimeSeed).nextInt(10) == 0; // Paper - All 
     }
 
     public boolean unload(boolean save) {
-- 
2.12.1.windows.1

