From 3248d91328e0db854ea869ccffd5e630fb2cee56 Mon Sep 17 00:00:00 2001
From: kashike <kasike@vq.lc>
Date: Wed, 27 Sep 2017 15:37:17 -0400
Subject: [PATCH] Expose server CommandMap


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 0b0a0fc..5bac3b1 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1641,6 +1641,7 @@ public final class CraftServer implements Server {
         return this.helpMap;
     }
 
+    @Override // Paper - add override
     public SimpleCommandMap getCommandMap() {
         return this.commandMap;
     }
-- 
2.14.1.windows.1

