From 046d9944e35da8aeaef665ad19b41b6c74f2f462 Mon Sep 17 00:00:00 2001
From: kashike <kashike@vq.lc>
Date: Wed, 27 Sep 2017 15:41:24 -0400
Subject: [PATCH] Be a bit more informative in maxHealth exception


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 4d83661..fdf9920 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -100,7 +100,10 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
                 this.getHandle().onDeath(DamageSource.GENERIC);
             }
         } else {
-            throw new IllegalArgumentException("Health must be between 0 and " + this.getMaxHealth() + "(" + health + ")");
+            // Paper - Be more informative
+            throw new IllegalArgumentException("Health must be between 0 and " + getMaxHealth() + ", but was " + health
+                    + ". (attribute base value: " + this.getHandle().getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).getBaseValue()
+                    + (this instanceof CraftPlayer ? ", player: " + this.getName() + ')' : ')'));
         }
     }
 
-- 
2.18.0

