From d09979d48937d7fd593e693cef8d849d2db432c7 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Wed, 27 Sep 2017 16:36:26 -0400
Subject: [PATCH] Vehicle Event Cancellation Changes


diff --git a/src/main/java/net/minecraft/entity/Entity.java b/src/main/java/net/minecraft/entity/Entity.java
index ebe0576..4cedc27 100644
--- a/src/main/java/net/minecraft/entity/Entity.java
+++ b/src/main/java/net/minecraft/entity/Entity.java
@@ -1949,6 +1949,7 @@ public abstract class Entity implements ICommandSender {
         if (passenger.getRidingEntity() == this) {
             throw new IllegalStateException("Use x.stopRiding(y), not y.removePassenger(x)");
         } else {
+            passenger.ridingEntity = this; // Paper - Set the vehicle back for the event
             CraftEntity craft = (CraftEntity) passenger.getBukkitEntity().getVehicle();
             Entity orig = craft == null ? null : craft.getHandle();
 
@@ -1963,7 +1964,12 @@ public abstract class Entity implements ICommandSender {
                 }
             }
 
-            Bukkit.getPluginManager().callEvent(new EntityDismountEvent(passenger.getBukkitEntity(), this.getBukkitEntity()));
+            // Paper start - make EntityDismountEvent cancellable
+            if (!new org.spigotmc.event.entity.EntityDismountEvent(passenger.getBukkitEntity(), this.getBukkitEntity()).callEvent()) {
+                return;
+            }
+            passenger.ridingEntity = null;
+            // Paper end
             this.riddenByEntities.remove(passenger);
             passenger.rideCooldown = 60;
         }
-- 
2.14.1

