From dc1bef3578345f573d46ea036dfba81b35af25c7 Mon Sep 17 00:00:00 2001
From: Joseph Hirschfeld <joe@ibj.io>
Date: Wed, 27 Sep 2017 17:31:58 -0400
Subject: [PATCH] Ensure inv drag is in bounds


diff --git a/src/main/java/net/minecraft/inventory/Container.java b/src/main/java/net/minecraft/inventory/Container.java
index 08402fa..1d1ee23 100644
--- a/src/main/java/net/minecraft/inventory/Container.java
+++ b/src/main/java/net/minecraft/inventory/Container.java
@@ -148,7 +148,7 @@ public abstract class Container {
                     this.resetDrag();
                 }
             } else if (this.dragEvent == 1) {
-                Slot slot = this.inventorySlots.get(slotId);
+                Slot slot = i1 < this.inventorySlots.size() ? this.inventorySlots.get(i1) : null; // Paper - Ensure drag in bounds
                 ItemStack itemstack1 = playerinventory.getItemStack();
 
                 if (slot != null
-- 
2.14.3

