From da94c87c19eccf8b2f7d5852196ec2be74a2cac8 Mon Sep 17 00:00:00 2001
From: Joseph Hirschfeld <joe@ibj.io>
Date: Thu, 3 Mar 2016 02:33:53 -0600
Subject: [PATCH] Ensure inv drag is in bounds


diff --git a/src/main/java/net/minecraft/inventory/Container.java b/src/main/java/net/minecraft/inventory/Container.java
index 5d4f493..7094b4f 100644
--- a/src/main/java/net/minecraft/inventory/Container.java
+++ b/src/main/java/net/minecraft/inventory/Container.java
@@ -163,7 +163,7 @@ public abstract class Container {
                     this.resetDrag();
                 }
             } else if (this.dragEvent == 1) {
-                Slot slot = (Slot)this.inventorySlots.get(slotId);
+                Slot slot = slotId < this.inventorySlots.size() ? (Slot)this.inventorySlots.get(slotId) : null; // Paper - Ensure drag in bounds
                 ItemStack itemstack1 = playerinventory.getItemStack();
 
                 if (slot != null && canAddItemToSlot(slot, itemstack1, true) && slot.isItemValid(itemstack1) && (this.dragMode == 2 || itemstack1.getCount() > this.dragSlots.size()) && this.canDragIntoSlot(slot)) {
-- 
2.12.1.windows.1

