From 58afb00fd39a9bfd25953786ef20c9a2ada2f9c1 Mon Sep 17 00:00:00 2001
From: Joseph Hirschfeld <joe@ibj.io>
Date: Sat, 30 Sep 2017 17:22:41 -0400
Subject: [PATCH] Add configurable portal search radius


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 05e0c0e..d904700 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -228,4 +228,9 @@ public class PaperWorldConfig {
     private void allChunksAreSlimeChunks() {
         allChunksAreSlimeChunks = getBoolean("all-chunks-are-slime-chunks", false);
     }
+
+    public int portalSearchRadius;
+    private void portalSearchRadius() {
+        portalSearchRadius = getInt("portal-search-radius", 128);
+    }
 }
diff --git a/src/main/java/net/minecraft/world/Teleporter.java b/src/main/java/net/minecraft/world/Teleporter.java
index 27ee629..e810539 100644
--- a/src/main/java/net/minecraft/world/Teleporter.java
+++ b/src/main/java/net/minecraft/world/Teleporter.java
@@ -96,7 +96,7 @@ public class Teleporter {
     }
 
     public boolean placeInExistingPortal(Entity entityIn, float rotationYaw) {
-        BlockPos found = this.findPortal(entityIn.posX, entityIn.posY, entityIn.posZ, 128);
+        BlockPos found = this.findPortal(entityIn.posX, entityIn.posY, entityIn.posZ, world.paperConfig.portalSearchRadius); // Paper - Configurable search radius
 
         if (found == null) {
             return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java b/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java
index 23733b6..0621e46 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java
@@ -8,7 +8,7 @@ import org.bukkit.TravelAgent;
 
 public class CraftTravelAgent extends Teleporter implements TravelAgent {
     public static TravelAgent DEFAULT = null;
-    private int searchRadius = 128;
+    private int searchRadius = world.paperConfig.portalSearchRadius; // Paper - Configurable search radius
     private int creationRadius = 16;
     private boolean canCreatePortal = true;
 
-- 
2.14.3

