From d878d6eaaea6aab05047f3bc88926e5081cf661e Mon Sep 17 00:00:00 2001
From: Joseph Hirschfeld <joe@ibj.io>
Date: Sat, 30 Sep 2017 16:46:53 -0400
Subject: [PATCH] Change implementation of (tile)entity removal list


diff --git a/src/main/java/net/minecraft/world/World.java b/src/main/java/net/minecraft/world/World.java
index c83282b..e7c2cc5 100644
--- a/src/main/java/net/minecraft/world/World.java
+++ b/src/main/java/net/minecraft/world/World.java
@@ -9,6 +9,8 @@ import com.google.common.collect.Maps;
 import java.util.*;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
+
+import com.google.common.collect.Sets;
 import net.minecraft.advancements.AdvancementManager;
 import net.minecraft.advancements.FunctionManager;
 import net.minecraft.block.Block;
@@ -112,11 +114,11 @@ public abstract class World implements IBlockAccess {
                     }
                 }
             };
-    protected final List<Entity> unloadedEntityList = Lists.<Entity>newArrayList();
+    protected final Set<Entity> unloadedEntityList = Sets.newHashSet(); // Paper
     public final List<TileEntity> loadedTileEntityList = Lists.<TileEntity>newArrayList();
     public final List<TileEntity> tickableTileEntities = Lists.<TileEntity>newArrayList();
     private final List<TileEntity> addedTileEntityList = Lists.<TileEntity>newArrayList();
-    private final List<TileEntity> tileEntitiesToBeRemoved = Lists.<TileEntity>newArrayList();
+    private final Set<TileEntity> tileEntitiesToBeRemoved = Sets.newHashSet(); // Paper
     public final List<EntityPlayer> playerEntities = Lists.<EntityPlayer>newArrayList();
     public final List<Entity> weatherEffects = Lists.<Entity>newArrayList();
     protected final IntHashMap<Entity> entitiesById = new IntHashMap<Entity>();
@@ -1511,8 +1513,8 @@ public abstract class World implements IBlockAccess {
         timings.entityRemoval.startTiming(); // Paper
         this.loadedEntityList.removeAll(this.unloadedEntityList);
 
-        for (int var20 = 0; var20 < this.unloadedEntityList.size(); ++var20) {
-            Entity entity = this.unloadedEntityList.get(var20);
+        // Paper start - Set based removal lists
+        for (Entity entity : unloadedEntityList) {
             int k = entity.chunkCoordX;
             int j = entity.chunkCoordZ;
 
@@ -1521,9 +1523,10 @@ public abstract class World implements IBlockAccess {
             }
         }
 
-        for (int var21 = 0; var21 < this.unloadedEntityList.size(); ++var21) {
-            this.onEntityRemoved(this.unloadedEntityList.get(var21));
+        for (Entity entity : unloadedEntityList) {
+            this.onEntityRemoved(entity);
         }
+        // Paper end
 
         this.unloadedEntityList.clear();
         this.tickPlayers();
-- 
2.14.1.windows.1

