From efeea2e950327f53cc9a43c3cc2eab765273fe6f Mon Sep 17 00:00:00 2001
From: Joseph Hirschfeld <joe@ibj.io>
Date: Sat, 30 Sep 2017 17:26:38 -0400
Subject: [PATCH] Add velociy warnings


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 5bac3b1..cc2438a 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -224,6 +224,7 @@ public final class CraftServer implements Server {
     private final List<CraftPlayer> playerView;
     public int reloadCount;
     private final Spigot spigot;
+    public static Exception excessiveVelEx; // Paper - Velocity warnings
 
     public CraftServer(MinecraftServer console, PlayerList playerList) {
         this.pluginManager = new SimplePluginManager(this, this.commandMap);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index c6681aa..76301fa 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -617,6 +617,13 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     public void setVelocity(Vector velocity) {
         Preconditions.checkArgument(velocity != null, "velocity");
         velocity.checkFinite();
+
+        // Paper start - Warn server owners when plugins try to set super high velocities
+        if (!(this instanceof org.bukkit.entity.Projectile) && isUnsafeVelocity(velocity)) {
+            CraftServer.excessiveVelEx = new Exception("Excessive velocity set detected: tried to set velocity of entity " + entity.getName() + " id #" + getEntityId() + " to (" + velocity.getX() + "," + velocity.getY() + "," + velocity.getZ() + ").");
+        }
+        // Paper end
+
         this.entity.motionX = velocity.getX();
         this.entity.motionY = velocity.getY();
         this.entity.motionZ = velocity.getZ();
@@ -1045,6 +1052,29 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         Location origin = getHandle().origin;
         return origin == null ? null : origin.clone();
     }
+
+    /**
+     * Checks if the given velocity is not necessarily safe in all situations.
+     * This function returning true does not mean the velocity is dangerous or to be avoided, only that it may be
+     * a detriment to performance on the server.
+     *
+     * It is not to be used as a hard rule of any sort.
+     * Paper only uses it to warn server owners in watchdog crashes.
+     *
+     * @param vel incoming velocity to check
+     * @return if the velocity has the potential to be a performance detriment
+     */
+    private static boolean isUnsafeVelocity(Vector vel) {
+        final double x = vel.getX();
+        final double y = vel.getY();
+        final double z = vel.getZ();
+
+        if (x > 4 || x < -4 || y > 4 || y < -4 || z > 4 || z < -4) {
+            return true;
+        }
+
+        return false;
+    }
     // Paper end
 
     public Spigot spigot() {
diff --git a/src/main/java/org/spigotmc/WatchdogThread.java b/src/main/java/org/spigotmc/WatchdogThread.java
index 2738156..d669c05 100644
--- a/src/main/java/org/spigotmc/WatchdogThread.java
+++ b/src/main/java/org/spigotmc/WatchdogThread.java
@@ -53,7 +53,17 @@ public class WatchdogThread extends Thread {
                     log.log(Level.SEVERE, "During the run of the server, a physics stackoverflow was supressed");
                     log.log(Level.SEVERE, "near " + World.blockLocation);
                 }
-
+                // Paper start - Warn in watchdog if an excessive velocity was ever set
+                if ( org.bukkit.craftbukkit.CraftServer.excessiveVelEx != null ) {
+                    log.log(Level.SEVERE, "------------------------------");
+                    log.log(Level.SEVERE, "During the run of the server, a plugin set an excessive velocity on an entity");
+                    log.log(Level.SEVERE, "This may be the cause of the issue, or it may be entirely unrelated");
+                    log.log(Level.SEVERE, org.bukkit.craftbukkit.CraftServer.excessiveVelEx.getMessage());
+                    for (StackTraceElement stack : org.bukkit.craftbukkit.CraftServer.excessiveVelEx.getStackTrace()) {
+                        log.log( Level.SEVERE, "\t\t" + stack );
+                    }
+                }
+                // Paper end
                 log.log(Level.SEVERE, "------------------------------");
                 log.log(Level.SEVERE, "Server thread dump (Look for plugins here before reporting to Paper!):");
                 dumpThread(
-- 
2.14.1.windows.1

