From 628cb132dd4203222486b5b4cad10c2ec92f863e Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sat, 30 Sep 2017 21:43:57 -0400
Subject: [PATCH] Add methods for working with arrows stuck in living entities


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index f1c12e1..1597005 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -553,4 +553,16 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     public boolean isCollidable() {
         return this.getHandle().collides;
     }
+
+    // Paper start
+    @Override
+    public int getArrowsStuck() {
+        return getHandle().getArrowCountInEntity();
+    }
+
+    @Override
+    public void setArrowsStuck(int arrows) {
+        getHandle().setArrowCountInEntity(arrows);
+    }
+    // Paper end
 }
-- 
2.14.1.windows.1

