From c250be704c835c602c09c7124a68d0f78b9ce8d8 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sat, 30 Sep 2017 21:43:57 -0400
Subject: [PATCH] Add methods for working with arrows stuck in living entities


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index fdf9920..1865d79 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -507,4 +507,16 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     public boolean isCollidable() {
         return this.getHandle().collides;
     }
+
+    // Paper start
+    @Override
+    public int getArrowsStuck() {
+        return getHandle().getArrowCountInEntity();
+    }
+
+    @Override
+    public void setArrowsStuck(int arrows) {
+        getHandle().setArrowCountInEntity(arrows);
+    }
+    // Paper end
 }
-- 
2.18.0

