From 03d9abfef57cf96a006ef8828cad14348ae99f3a Mon Sep 17 00:00:00 2001
From: William <admin@domnian.com>
Date: Sat, 30 Sep 2017 22:45:22 -0400
Subject: [PATCH] Allow Reloading of Custom Permissions

https://github.com/PaperMC/Paper/issues/49

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 4a09904..1e53b2c 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1887,5 +1887,20 @@ public final class CraftServer implements Server {
                 console.tps15.getAverage()
         };
     }
+
+    @Override
+    public void reloadPermissions() {
+        ((SimplePluginManager) pluginManager).clearPermissions();
+        loadCustomPermissions();
+        for (Plugin plugin : pluginManager.getPlugins()) {
+            plugin.getDescription().getPermissions().forEach((perm) -> {
+                try {
+                    pluginManager.addPermission(perm);
+                } catch (IllegalArgumentException ex) {
+                    getLogger().log(Level.WARNING, "Plugin " + plugin.getDescription().getFullName() + " tried to register permission '" + perm.getName() + "' but it's already registered", ex);
+                }
+            });
+        }
+    }
     // Paper end
 }
-- 
2.18.0

