From 752a5df0d2b7d36a6ff9da9da841350bead28eb3 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 1 Oct 2017 00:17:42 -0400
Subject: [PATCH] Catch Async PlayerChunkMap operations


diff --git a/src/main/java/net/minecraft/server/management/PlayerChunkMap.java b/src/main/java/net/minecraft/server/management/PlayerChunkMap.java
index 3eddccd..20a1fca 100644
--- a/src/main/java/net/minecraft/server/management/PlayerChunkMap.java
+++ b/src/main/java/net/minecraft/server/management/PlayerChunkMap.java
@@ -465,10 +465,12 @@ public class PlayerChunkMap {
     }
 
     public void entryChanged(PlayerChunkMapEntry entry) {
+        org.spigotmc.AsyncCatcher.catchOp("Async Player Chunk Add"); // Paper
         this.dirtyEntries.add(entry);
     }
 
     public void removeEntry(PlayerChunkMapEntry entry) {
+        org.spigotmc.AsyncCatcher.catchOp("Async Player Chunk Remove"); // Paper
         ChunkPos chunkcoordintpair = entry.getPos();
         long i = getIndex(chunkcoordintpair.x, chunkcoordintpair.z);
         entry.updateChunkInhabitedTime();
-- 
2.14.1.windows.1

