From df920c4f95f73d7093f53c4a7e75e6a91c992a89 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 1 Oct 2017 01:15:37 -0400
Subject: [PATCH] Entity AddTo/RemoveFrom World Events


diff --git a/src/main/java/net/minecraft/world/World.java b/src/main/java/net/minecraft/world/World.java
index 0cff92f..0fbb75f 100644
--- a/src/main/java/net/minecraft/world/World.java
+++ b/src/main/java/net/minecraft/world/World.java
@@ -1205,6 +1205,7 @@ public abstract class World implements IBlockAccess {
         }
 
         entityIn.valid = true;
+        new com.destroystokyo.paper.event.entity.EntityAddToWorldEvent(entityIn.getBukkitEntity()).callEvent(); // Paper - fire while valid
     }
 
     protected void onEntityRemoved(Entity entityIn) {
@@ -1212,6 +1213,7 @@ public abstract class World implements IBlockAccess {
             ((IWorldEventListener) this.eventListeners.get(i)).onEntityRemoved(entityIn);
         }
 
+        new com.destroystokyo.paper.event.entity.EntityRemoveFromWorldEvent(entityIn.getBukkitEntity()).callEvent(); // Paper - fire while valid
         entityIn.valid = false;
     }
 
-- 
2.14.3

