From 9be47e6eef49aa4216b9fe2c285c08c4c3f714c5 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 1 Oct 2017 01:15:37 -0400
Subject: [PATCH] Entity AddTo/RemoveFrom World Events


diff --git a/src/main/java/net/minecraft/world/World.java b/src/main/java/net/minecraft/world/World.java
index a3a4a71..0f23cd7 100644
--- a/src/main/java/net/minecraft/world/World.java
+++ b/src/main/java/net/minecraft/world/World.java
@@ -1166,6 +1166,7 @@ public abstract class World implements IBlockAccess {
         }
 
         entityIn.valid = true;
+        new com.destroystokyo.paper.event.entity.EntityAddToWorldEvent(entityIn.getBukkitEntity()).callEvent(); // Paper - fire while valid
     }
 
     protected void onEntityRemoved(Entity entityIn) {
@@ -1173,6 +1174,7 @@ public abstract class World implements IBlockAccess {
             ((IWorldEventListener) this.eventListeners.get(i)).onEntityRemoved(entityIn);
         }
 
+        new com.destroystokyo.paper.event.entity.EntityRemoveFromWorldEvent(entityIn.getBukkitEntity()).callEvent(); // Paper - fire while valid
         entityIn.valid = false;
     }
 
-- 
2.18.0

