From 472bc5828f9507e1a29ba75df209b2d85dcc4bf6 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 7 Oct 2017 21:08:48 -0400
Subject: [PATCH] Add PlayerUseUnknownEntityEvent


diff --git a/src/main/java/net/minecraft/network/NetHandlerPlayServer.java b/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
index d743b6e..2184ec1 100644
--- a/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
+++ b/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
@@ -1801,6 +1801,17 @@ public class NetHandlerPlayServer implements INetHandlerPlayServer, ITickable {
                         }
                     }
                 }
+
+                    // Paper start - fire event
+                    else {
+                        this.server.getPluginManager().callEvent(new com.destroystokyo.paper.event.player.PlayerUseUnknownEntityEvent(
+                                this.getPlayer(),
+                                packetIn.entityId,
+                                packetIn.getAction() == CPacketUseEntity.Action.ATTACK,
+                                packetIn.getHand() == EnumHand.MAIN_HAND ? EquipmentSlot.HAND : EquipmentSlot.OFF_HAND
+                        ));
+                    }
+                    // Paper end
             }
         }
     }
-- 
2.18.0

