From 044d87004e7e7810ddda2e8c0766a70c55a710bc Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 7 Oct 2017 21:26:45 -0400
Subject: [PATCH] Configurable Grass Spread Tick Rate


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 09e9fe4..06be217 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -272,4 +272,10 @@ public class PaperWorldConfig {
     private void useInhabitedTime() {
         useInhabitedTime = getBoolean("use-chunk-inhabited-timer", true);
     }
+
+    public int grassUpdateRate = 1;
+    private void grassUpdateRate() {
+        grassUpdateRate = Math.max(0, getInt("grass-spread-tick-rate", grassUpdateRate));
+        log("Grass Spread Tick Rate: " + grassUpdateRate);
+    }
 }
diff --git a/src/main/java/net/minecraft/block/BlockGrass.java b/src/main/java/net/minecraft/block/BlockGrass.java
index 9dcf66f..b1eba4f 100644
--- a/src/main/java/net/minecraft/block/BlockGrass.java
+++ b/src/main/java/net/minecraft/block/BlockGrass.java
@@ -9,6 +9,7 @@ import net.minecraft.block.state.IBlockState;
 import net.minecraft.creativetab.CreativeTabs;
 import net.minecraft.init.Blocks;
 import net.minecraft.item.Item;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
@@ -34,6 +35,7 @@ public class BlockGrass extends Block implements IGrowable {
     }
 
     public void updateTick(World worldIn, BlockPos pos, IBlockState state, Random rand) {
+        if (worldIn.paperConfig.grassUpdateRate != 1 && (worldIn.paperConfig.grassUpdateRate < 1 || (MinecraftServer.currentTick + pos.hashCode()) % worldIn.paperConfig.grassUpdateRate != 0)) { return; } // Paper
         if (!worldIn.isRemote) {
             int lightLevel = -1; // Paper
             if (worldIn.getBlockState(pos.up()).getLightOpacity() > 2 && (lightLevel = worldIn.getLightFromNeighbors(pos.up())) < 4) { // Paper - move light check to end to avoid unneeded light lookups
-- 
2.18.0

