From ec42f82b0bf4e080ba77bf6711488964c73bbb67 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 7 Oct 2017 21:38:10 -0400
Subject: [PATCH] Don't spam reload spawn chunks in nether/end


diff --git a/src/main/java/net/minecraft/world/WorldProvider.java b/src/main/java/net/minecraft/world/WorldProvider.java
index df4b60e..2886a14 100644
--- a/src/main/java/net/minecraft/world/WorldProvider.java
+++ b/src/main/java/net/minecraft/world/WorldProvider.java
@@ -155,6 +155,6 @@ public abstract class WorldProvider {
     public void onWorldUpdateEntities() {}
 
     public boolean canDropChunk(int x, int z) {
-        return true;
+        return !world.isSpawnChunk(x, z); // Paper - Use isSpawnChunk for all worlds
     }
 }
-- 
2.18.0

