From bcb492fff1ad245326e9ffbdcea86c3a3b8e4888 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sat, 7 Oct 2017 22:28:52 -0400
Subject: [PATCH] Workaround for setting passengers on players

SPIGOT-1915 & GH-114

diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index de681bd..e95d228 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1596,6 +1596,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.getHandle().affectsSpawning = affects;
     }
 
+    // Paper start - Ugly workaround for SPIGOT-1915 & GH-114
+    @Override
+    public boolean setPassenger(org.bukkit.entity.Entity passenger) {
+        final boolean wasSet = super.setPassenger(passenger);
+        if (wasSet) {
+            this.getHandle().connection.sendPacket(new net.minecraft.network.play.server.SPacketSetPassengers(this.getHandle()));
+        }
+        return wasSet;
+    }
+    // Paper end
+
     @Override
     public boolean getAffectsSpawning() {
         return this.getHandle().affectsSpawning;
-- 
2.14.3.windows.1

