From 2549cb227bca0a1846201b4756da77a632ac48f6 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 7 Oct 2017 22:34:59 -0400
Subject: [PATCH] Don't tick Skulls - unused code


diff --git a/src/main/java/net/minecraft/tileentity/TileEntitySkull.java b/src/main/java/net/minecraft/tileentity/TileEntitySkull.java
index 6bfcbaa..99dcf70 100644
--- a/src/main/java/net/minecraft/tileentity/TileEntitySkull.java
+++ b/src/main/java/net/minecraft/tileentity/TileEntitySkull.java
@@ -33,7 +33,7 @@ import net.minecraft.util.Mirror;
 import net.minecraft.util.Rotation;
 import net.minecraft.util.StringUtils;
 
-public class TileEntitySkull extends TileEntity implements ITickable {
+public class TileEntitySkull extends TileEntity { // Paper - Remove ITickable
     private int skullType;
     public int skullRotation;
     private GameProfile playerProfile;
-- 
2.18.0

