From 029f292c45d3768e465d015832361141778aa000 Mon Sep 17 00:00:00 2001
From: Isaac Moore <rmsy@me.com>
Date: Sun, 8 Oct 2017 00:08:08 -0400
Subject: [PATCH] Implement PlayerLocaleChangeEvent


diff --git a/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java b/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java
index 82ade90..d13cc0a 100644
--- a/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java
+++ b/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java
@@ -140,7 +140,7 @@ import org.bukkit.inventory.MainHand;
 
 public class EntityPlayerMP extends EntityPlayer implements IContainerListener {
     private static final Logger LOGGER = LogManager.getLogger();
-    public String language = "en_us";
+    public String language = null; // Paper - Default to null
     public NetHandlerPlayServer connection;
     public final MinecraftServer mcServer;
     public final PlayerInteractionManager interactionManager;
@@ -1311,12 +1311,23 @@ public class EntityPlayerMP extends EntityPlayer implements IContainerListener {
             this.mcServer.server.getPluginManager().callEvent(event);
         }
 
-        if (!this.language.equals(packetIn.getLang())) {
+        // Paper start - add PlayerLocaleChangeEvent
+        // Since the field is initialized to null, this event should always fire the first time the packet is received
+        String oldLanguage = this.language;
+        this.language = packetIn.getLang();
+        if (!this.language.equals(oldLanguage)) {
+            new com.destroystokyo.paper.event.player.PlayerLocaleChangeEvent(this.getBukkitEntity(), oldLanguage, this.language).callEvent();
+        }
+
+        // Compat with Bukkit
+        oldLanguage = oldLanguage != null ? oldLanguage : "en_us";
+        // Paper end
+
+        if (!oldLanguage.equals(packetIn.getLang())) {
             PlayerLocaleChangeEvent event = new PlayerLocaleChangeEvent(this.getBukkitEntity(), packetIn.getLang());
             this.mcServer.server.getPluginManager().callEvent(event);
         }
 
-        this.language = packetIn.getLang();
         this.chatVisibility = packetIn.getChatVisibility();
         this.chatColours = packetIn.isColorsEnabled();
         this.getDataManager().set(PLAYER_MODEL_FLAG, Byte.valueOf((byte) packetIn.getModelPartFlags()));
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index ea18f09..323cc58 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -242,7 +242,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
                 }
 
                 public String getLocale() {
-                    return CraftPlayer.this.getHandle().language;
+                    return CraftPlayer.this.getLocale(); // Paper
                 }
 
                 public Set<Player> getHiddenPlayers() {
@@ -1630,7 +1630,10 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
 
     public String getLocale() {
-        return this.getHandle().language;
+        // Paper start - Locale change event
+        final String locale = getHandle().language;
+        return locale != null ? locale : "en_us";
+        // Paper end
     }
 
     // Paper start
-- 
2.18.0

