From be596678fe50a3f5ef41061fe227f80fd817b425 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sun, 8 Oct 2017 00:30:17 -0400
Subject: [PATCH] SPIGOT-1401: Fix dispenser, dropper, furnace placement


diff --git a/src/main/java/net/minecraft/block/BlockDispenser.java b/src/main/java/net/minecraft/block/BlockDispenser.java
index a301853..d640fa2 100644
--- a/src/main/java/net/minecraft/block/BlockDispenser.java
+++ b/src/main/java/net/minecraft/block/BlockDispenser.java
@@ -51,6 +51,9 @@ public class BlockDispenser extends BlockContainer {
         return 4;
     }
 
+    // Paper start - Removed override of onBlockAdded that was reversing placement direction when
+    // adjacent to another block, which was not consistent with single player block placement
+    /*
     public void onBlockAdded(World worldIn, BlockPos pos, IBlockState state) {
         super.onBlockAdded(worldIn, pos, state);
         this.setDefaultDirection(worldIn, pos, state);
@@ -80,6 +83,8 @@ public class BlockDispenser extends BlockContainer {
             worldIn.setBlockState(pos, state.withProperty(FACING, enumdirection).withProperty(TRIGGERED, Boolean.valueOf(false)), 2);
         }
     }
+    */
+    // Paper end
 
     public boolean onBlockActivated(
             World worldIn,
diff --git a/src/main/java/net/minecraft/block/BlockFurnace.java b/src/main/java/net/minecraft/block/BlockFurnace.java
index 2cada96..9a35a1b 100644
--- a/src/main/java/net/minecraft/block/BlockFurnace.java
+++ b/src/main/java/net/minecraft/block/BlockFurnace.java
@@ -39,6 +39,9 @@ public class BlockFurnace extends BlockContainer {
         return Item.getItemFromBlock(Blocks.FURNACE);
     }
 
+    // Paper start - Removed override of onBlockAdded that was reversing placement direction when
+    // adjacent to another block, which was not consistent with single player block placement
+    /*
     public void onBlockAdded(World worldIn, BlockPos pos, IBlockState state) {
         this.setDefaultFacing(worldIn, pos, state);
     }
@@ -64,6 +67,8 @@ public class BlockFurnace extends BlockContainer {
             worldIn.setBlockState(pos, state.withProperty(FACING, var8), 2);
         }
     }
+    */
+    // Paper end
 
     public boolean onBlockActivated(
             World worldIn,
-- 
2.14.3

