From 66b08cdf6ad9595339c9813821fe9b36ff3715cc Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 9 Oct 2017 17:25:14 -0400
Subject: [PATCH] Do not load chunks for pathfinding


diff --git a/src/main/java/net/minecraft/world/ChunkCache.java b/src/main/java/net/minecraft/world/ChunkCache.java
index 7f91929..0cf4922 100644
--- a/src/main/java/net/minecraft/world/ChunkCache.java
+++ b/src/main/java/net/minecraft/world/ChunkCache.java
@@ -27,7 +27,7 @@ public class ChunkCache implements IBlockAccess {
 
         for (int var7 = this.chunkX; var7 <= var5; ++var7) {
             for (int var8 = this.chunkZ; var8 <= var6; ++var8) {
-                this.chunkArray[var7 - this.chunkX][var8 - this.chunkZ] = worldIn.getChunkFromChunkCoords(var7, var8);
+                this.chunkArray[var7 - this.chunkX][var8 - this.chunkZ] = worldIn.getChunkIfLoaded(var7, var8); // Paper - Do not load chunks for pathfinding
             }
         }
 
-- 
2.14.1.windows.1

