From 302ab82f96fa4ab5a56760440a1d4f0c39460c02 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Mon, 9 Oct 2017 18:18:20 -0400
Subject: [PATCH] System property for disabling watchdoge


diff --git a/src/main/java/org/spigotmc/WatchdogThread.java b/src/main/java/org/spigotmc/WatchdogThread.java
index d669c05..07ce350 100644
--- a/src/main/java/org/spigotmc/WatchdogThread.java
+++ b/src/main/java/org/spigotmc/WatchdogThread.java
@@ -41,7 +41,7 @@ public class WatchdogThread extends Thread {
 
     public void run() {
         while (!this.stopping) {
-            if (this.lastTick != 0L && System.currentTimeMillis() > this.lastTick + this.timeoutTime) {
+            if (this.lastTick != 0L && System.currentTimeMillis() > this.lastTick + this.timeoutTime && !Boolean.getBoolean("disable.watchdog")) { // Paper - Add property to disable
                 Logger log = Bukkit.getServer().getLogger();
                 log.log(Level.SEVERE, "The server has stopped responding!");
                 log.log(Level.SEVERE, "Please report this to https://aquifermc.org/");
-- 
2.18.0

