From 8e9169d0e4e87067d6544cb94da4b0bfe261806c Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 9 Oct 2017 18:21:03 -0400
Subject: [PATCH] Optimize EAR


diff --git a/src/main/java/org/spigotmc/ActivationRange.java b/src/main/java/org/spigotmc/ActivationRange.java
index ce0f63c..260d202 100644
--- a/src/main/java/org/spigotmc/ActivationRange.java
+++ b/src/main/java/org/spigotmc/ActivationRange.java
@@ -1,6 +1,8 @@
 package org.spigotmc;
 
 import java.util.List;
+
+import com.destroystokyo.paper.util.MCUtil;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityCreature;
 import net.minecraft.entity.EntityLivingBase;
@@ -70,6 +72,7 @@ public class ActivationRange {
         maxRange = Math.max(maxRange, miscActivationRange);
         maxRange = Math.min((world.spigotConfig.viewDistance << 4) - 8, maxRange);
 
+        Chunk chunk; // Paper
         for (EntityPlayer player : world.playerEntities) {
             player.activatedTick = (long) MinecraftServer.currentTick;
             maxBB = player.getEntityBoundingBox().grow((double) maxRange, 256.0D, (double) maxRange);
@@ -83,8 +86,8 @@ public class ActivationRange {
 
             for (int i1 = i; i1 <= j; ++i1) {
                 for (int j1 = k; j1 <= l; ++j1) {
-                    if (world.getWorld().isChunkLoaded(i1, j1)) {
-                        activateChunkEntities(world.getChunkFromChunkCoords(i1, j1));
+                    if ((chunk = MCUtil.getLoadedChunkWithoutMarkingActive(world, i1, j1 )) != null) { // Paper
+                        activateChunkEntities(chunk); // Paper
                     }
                 }
             }
-- 
2.18.0

