From fed410ddfde3e4fc750bb5febd8538f572833e5a Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 9 Oct 2017 18:50:13 -0400
Subject: [PATCH] Only send Dragon/Wither Death sounds to same world

Also fix view distance lookup

diff --git a/src/main/java/net/minecraft/entity/boss/EntityDragon.java b/src/main/java/net/minecraft/entity/boss/EntityDragon.java
index f12ca4d..9217665 100644
--- a/src/main/java/net/minecraft/entity/boss/EntityDragon.java
+++ b/src/main/java/net/minecraft/entity/boss/EntityDragon.java
@@ -684,9 +684,11 @@ public class EntityDragon extends EntityLiving implements IEntityMultiPart, IMob
             }
 
             if (this.deathTicks == 1) {
-                int viewDistance = ((WorldServer) this.world).getServer().getViewDistance() * 16;
-
-                for (EntityPlayerMP player : MinecraftServer.getServer().getPlayerList().playerEntityList) {
+                // Paper start
+                for (EntityPlayer human : world.playerEntities) {
+                    EntityPlayerMP player = (EntityPlayerMP) human;
+                    int viewDistance = player.getViewDistance();
+                    // Paper end
                     double deltaX = this.posX - player.posX;
                     double deltaZ = this.posZ - player.posZ;
                     double distanceSquared = deltaX * deltaX + deltaZ * deltaZ;
diff --git a/src/main/java/net/minecraft/entity/boss/EntityWither.java b/src/main/java/net/minecraft/entity/boss/EntityWither.java
index 5b8dae7..1be384c 100644
--- a/src/main/java/net/minecraft/entity/boss/EntityWither.java
+++ b/src/main/java/net/minecraft/entity/boss/EntityWither.java
@@ -269,9 +269,11 @@ public class EntityWither extends EntityMob implements IRangedAttackMob {
                             this.world.getGameRules().getBoolean("mobGriefing"));
                 }
 
-                int viewDistance = ((WorldServer) this.world).getServer().getViewDistance() * 16;
-
-                for (EntityPlayerMP player : MinecraftServer.getServer().getPlayerList().playerEntityList) {
+                // Paper start
+                for (EntityPlayer human : world.playerEntities) {
+                    EntityPlayerMP player = (EntityPlayerMP) human;
+                    int viewDistance = player.getViewDistance();
+                    // Paper end
                     double deltaX = this.posX - player.posX;
                     double deltaZ = this.posZ - player.posZ;
                     double distanceSquared = deltaX * deltaX + deltaZ * deltaZ;
-- 
2.18.0

