From 710f780a90dcd10a3e46ffa00382f97b1ba0f788 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Tue, 10 Oct 2017 15:40:15 -0400
Subject: [PATCH] More informative vehicle moved wrongly message


diff --git a/src/main/java/net/minecraft/network/NetHandlerPlayServer.java b/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
index c80ccf7..73d2f15 100644
--- a/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
+++ b/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
@@ -533,7 +533,7 @@ public class NetHandlerPlayServer implements INetHandlerPlayServer, ITickable {
 
                 if (d10 > SpigotConfig.movedWronglyThreshold) {
                     flag1 = true;
-                    LOGGER.warn("{} moved wrongly!", (Object) entity.getName());
+                    LOGGER.warn(entity.getName() + " (vehicle of " + this.player.getName() + ") moved wrongly!"); // Paper - More informative
                 }
 
                 entity.setPositionAndRotation(d3, d4, d5, f, f1);
-- 
2.18.0

