From 36c52940ffb0bb0bd82c15ed5bef4e8faaef8045 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 10 Oct 2017 16:35:35 -0400
Subject: [PATCH] Remove FishingHook reference on Craft Entity removal


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java
index e76d3fb..ec65557 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java
@@ -58,6 +58,16 @@ public class CraftFish extends AbstractProjectile implements Fish {
         this.biteChance = chance;
     }
 
+    // Paper start
+    @Override
+    public void remove() {
+        super.remove();
+        if (getHandle().angler != null) {
+            getHandle().angler.fishEntity = null;
+        }
+    }
+    // Paper end
+
     @Deprecated
     public LivingEntity _INVALID_getShooter() {
         return (LivingEntity) this.getShooter();
-- 
2.14.2.windows.2

