From 82df8da574f0d5625c53c0ecd3cf98dae94ed72c Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 10 Oct 2017 16:35:35 -0400
Subject: [PATCH] Remove FishingHook reference on Craft Entity removal


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java
index 81a1003..dae97ab 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java
@@ -55,4 +55,14 @@ public class CraftFish extends AbstractProjectile implements Fish {
         Validate.isTrue(chance >= 0.0D && chance <= 1.0D, "The bite chance must be between 0 and 1.");
         this.biteChance = chance;
     }
+
+    // Paper start
+    @Override
+    public void remove() {
+        super.remove();
+        if (getHandle().angler != null) {
+            getHandle().angler.fishEntity = null;
+        }
+    }
+    // Paper end
 }
-- 
2.18.0

