From 2c0de3fc52764927833c3d9ac0bd6a493158e128 Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@outlook.com>
Date: Fri, 20 Oct 2017 19:29:09 -0400
Subject: [PATCH] Add ProjectileCollideEvent


diff --git a/src/main/java/net/minecraft/entity/projectile/EntityArrow.java b/src/main/java/net/minecraft/entity/projectile/EntityArrow.java
index a0646ee..7963336 100644
--- a/src/main/java/net/minecraft/entity/projectile/EntityArrow.java
+++ b/src/main/java/net/minecraft/entity/projectile/EntityArrow.java
@@ -221,6 +221,15 @@ public abstract class EntityArrow extends Entity implements IProjectile {
                 }
             }
 
+            // Paper start - Call ProjectileCollideEvent
+            if (movingobjectposition != null && movingobjectposition.entityHit != null) {
+                com.destroystokyo.paper.event.entity.ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition);
+                if (event.isCancelled()) {
+                    movingobjectposition = null;
+                }
+            }
+            // Paper end
+
             if (movingobjectposition != null) {
                 this.onHit(movingobjectposition);
             }
diff --git a/src/main/java/net/minecraft/entity/projectile/EntityFireball.java b/src/main/java/net/minecraft/entity/projectile/EntityFireball.java
index 6fa0c11..899a778 100644
--- a/src/main/java/net/minecraft/entity/projectile/EntityFireball.java
+++ b/src/main/java/net/minecraft/entity/projectile/EntityFireball.java
@@ -79,6 +79,15 @@ public abstract class EntityFireball extends Entity {
             ++this.ticksInAir;
             RayTraceResult movingobjectposition = ProjectileHelper.forwardsRaycast(this, true, this.ticksInAir >= 25, this.shootingEntity);
 
+            // Paper start - Call ProjectileCollideEvent
+            if (movingobjectposition != null && movingobjectposition.entityHit != null) {
+                com.destroystokyo.paper.event.entity.ProjectileCollideEvent event = CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition);
+                if (event.isCancelled()) {
+                    movingobjectposition = null;
+                }
+            }
+            // Paper end
+
             if (movingobjectposition != null) {
                 this.onImpact(movingobjectposition);
 
diff --git a/src/main/java/net/minecraft/entity/projectile/EntityFishHook.java b/src/main/java/net/minecraft/entity/projectile/EntityFishHook.java
index e86b45e..4f18929 100644
--- a/src/main/java/net/minecraft/entity/projectile/EntityFishHook.java
+++ b/src/main/java/net/minecraft/entity/projectile/EntityFishHook.java
@@ -263,6 +263,15 @@ public class EntityFishHook extends Entity {
         vec3d = new Vec3d(this.posX, this.posY, this.posZ);
         vec3d1 = new Vec3d(this.posX + this.motionX, this.posY + this.motionY, this.posZ + this.motionZ);
 
+        // Paper start - Call ProjectileCollideEvent
+        if (movingobjectposition != null && movingobjectposition.entityHit != null) {
+            com.destroystokyo.paper.event.entity.ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition);
+            if (event.isCancelled()) {
+                movingobjectposition = null;
+            }
+        }
+        // Paper end
+
         if (movingobjectposition != null) {
             vec3d1 = new Vec3d(movingobjectposition.hitVec.x, movingobjectposition.hitVec.y, movingobjectposition.hitVec.z);
         }
diff --git a/src/main/java/net/minecraft/entity/projectile/EntityLlamaSpit.java b/src/main/java/net/minecraft/entity/projectile/EntityLlamaSpit.java
index d19e0ce..37adc27 100644
--- a/src/main/java/net/minecraft/entity/projectile/EntityLlamaSpit.java
+++ b/src/main/java/net/minecraft/entity/projectile/EntityLlamaSpit.java
@@ -50,6 +50,15 @@ public class EntityLlamaSpit extends Entity implements IProjectile {
         vec3d = new Vec3d(this.posX, this.posY, this.posZ);
         vec3d1 = new Vec3d(this.posX + this.motionX, this.posY + this.motionY, this.posZ + this.motionZ);
 
+        // Paper start - Call ProjectileCollideEvent
+        if (movingobjectposition != null && movingobjectposition.entityHit != null) {
+            com.destroystokyo.paper.event.entity.ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition);
+            if (event.isCancelled()) {
+                movingobjectposition = null;
+            }
+        }
+        // Paper end
+
         if (movingobjectposition != null) {
             vec3d1 = new Vec3d(movingobjectposition.hitVec.x, movingobjectposition.hitVec.y, movingobjectposition.hitVec.z);
         }
diff --git a/src/main/java/net/minecraft/entity/projectile/EntityShulkerBullet.java b/src/main/java/net/minecraft/entity/projectile/EntityShulkerBullet.java
index 6cce682..32a297f 100644
--- a/src/main/java/net/minecraft/entity/projectile/EntityShulkerBullet.java
+++ b/src/main/java/net/minecraft/entity/projectile/EntityShulkerBullet.java
@@ -272,6 +272,15 @@ public class EntityShulkerBullet extends Entity {
 
                 RayTraceResult movingobjectposition = ProjectileHelper.forwardsRaycast(this, true, false, this.owner);
 
+                // Paper start - Call ProjectileCollideEvent
+                if (movingobjectposition != null && movingobjectposition.entityHit != null) {
+                    com.destroystokyo.paper.event.entity.ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition);
+                    if (event.isCancelled()) {
+                        movingobjectposition = null;
+                    }
+                }
+                // Paper end
+
                 if (movingobjectposition != null) {
                     this.bulletHit(movingobjectposition);
                 }
diff --git a/src/main/java/net/minecraft/entity/projectile/EntityThrowable.java b/src/main/java/net/minecraft/entity/projectile/EntityThrowable.java
index 34dcc08..8bd289c 100644
--- a/src/main/java/net/minecraft/entity/projectile/EntityThrowable.java
+++ b/src/main/java/net/minecraft/entity/projectile/EntityThrowable.java
@@ -180,6 +180,15 @@ public abstract class EntityThrowable extends Entity implements IProjectile {
             movingobjectposition = new RayTraceResult(entity);
         }
 
+        // Paper start - Call ProjectileCollideEvent
+        if (movingobjectposition != null && movingobjectposition.entityHit != null) {
+            com.destroystokyo.paper.event.entity.ProjectileCollideEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileCollideEvent(this, movingobjectposition);
+            if (event.isCancelled()) {
+                movingobjectposition = null;
+            }
+        }
+        // Paper end
+
         if (movingobjectposition != null) {
             if (movingobjectposition.typeOfHit == RayTraceResult.Type.BLOCK
                     && this.world.getBlockState(movingobjectposition.getBlockPos()).getBlock() == Blocks.PORTAL) {
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index bd92f0d..b666b2e 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -958,6 +958,16 @@ public class CraftEventFactory {
         return CraftItemStack.asNMSCopy(bitem);
     }
 
+    // Paper start
+    public static com.destroystokyo.paper.event.entity.ProjectileCollideEvent callProjectileCollideEvent(Entity entity, RayTraceResult position) {
+        Projectile projectile = (Projectile) entity.getBukkitEntity();
+        org.bukkit.entity.Entity collided = position.entityHit.getBukkitEntity();
+        com.destroystokyo.paper.event.entity.ProjectileCollideEvent event = new com.destroystokyo.paper.event.entity.ProjectileCollideEvent(projectile, collided);
+        Bukkit.getPluginManager().callEvent(event);
+        return event;
+    }
+    // Paper end
+
     public static ProjectileLaunchEvent callProjectileLaunchEvent(Entity entity) {
         Projectile bukkitEntity = (Projectile) entity.getBukkitEntity();
         ProjectileLaunchEvent event = new ProjectileLaunchEvent(bukkitEntity);
-- 
2.18.0

