From d5adaac9d40146da5293c4dfdc09fab09c906a01 Mon Sep 17 00:00:00 2001
From: kashike <kashike@vq.lc>
Date: Fri, 20 Oct 2017 20:06:52 -0400
Subject: [PATCH] Add API methods to control if armour stands can move


diff --git a/src/main/java/net/minecraft/entity/item/EntityArmorStand.java b/src/main/java/net/minecraft/entity/item/EntityArmorStand.java
index 4d7c03d..1ab5671 100644
--- a/src/main/java/net/minecraft/entity/item/EntityArmorStand.java
+++ b/src/main/java/net/minecraft/entity/item/EntityArmorStand.java
@@ -8,6 +8,7 @@ import net.minecraft.block.material.EnumPushReaction;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.entity.EntityLivingBase;
+import net.minecraft.entity.MoverType;
 import net.minecraft.entity.effect.EntityLightningBolt;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.projectile.EntityArrow;
@@ -85,6 +86,7 @@ public class EntityArmorStand extends EntityLivingBase {
     public Rotations rightArmRotation;
     public Rotations leftLegRotation;
     public Rotations rightLegRotation;
+    public boolean canMove = true; // Paper
 
     public EntityArmorStand(World worldIn) {
         super(worldIn);
@@ -784,4 +786,13 @@ public class EntityArmorStand extends EntityLivingBase {
     public boolean attackable() {
         return false;
     }
+
+    // Paper start
+    @Override
+    public void move(MoverType moveType, double x, double y, double z) {
+        if (this.canMove) {
+            super.move(moveType, x, y, z);
+        }
+    }
+    // Paper end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftArmorStand.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftArmorStand.java
index 183bae3..5477878 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftArmorStand.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftArmorStand.java
@@ -165,4 +165,14 @@ public class CraftArmorStand extends CraftLivingEntity implements ArmorStand {
     public void setMarker(boolean marker) {
         this.getHandle().setMarker(marker);
     }
+
+    @Override
+    public boolean canMove() {
+        return getHandle().canMove;
+    }
+
+    @Override
+    public void setCanMove(boolean move) {
+        getHandle().canMove = move;
+    }
 }
-- 
2.18.0

