From 9a2841928ae721e5d195255e5469f51469a65117 Mon Sep 17 00:00:00 2001
From: Alfie Cleveland <alfeh@me.com>
Date: Fri, 20 Oct 2017 20:19:22 -0400
Subject: [PATCH] Stricter isDisconnected and isMovementBlocked checks

Credit to prplz for figuring out specifics

diff --git a/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java b/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java
index 58c2b1a..7c76846 100644
--- a/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java
+++ b/src/main/java/net/minecraft/entity/player/EntityPlayerMP.java
@@ -1521,7 +1521,7 @@ public class EntityPlayerMP extends EntityPlayer implements IContainerListener {
     }
 
     public boolean isMovementBlocked() {
-        return super.isMovementBlocked() || !this.getBukkitEntity().isOnline();
+        return super.isMovementBlocked() || (this.connection != null && this.connection.isDisconnected()); // Paper - Stricter check
     }
 
     public Scoreboard getWorldScoreboard() {
diff --git a/src/main/java/net/minecraft/network/NetHandlerPlayServer.java b/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
index e342e2d..9eead06 100644
--- a/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
+++ b/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
@@ -2830,6 +2830,6 @@ public class NetHandlerPlayServer implements INetHandlerPlayServer, ITickable {
     }
 
     public final boolean isDisconnected() {
-        return !this.player.joining && !this.netManager.isChannelOpen();
+        return (!this.player.joining && !this.netManager.isChannelOpen()) || this.processedDisconnect; // Paper - Stricter check
     }
 }
-- 
2.18.0

