From a1a6c6fad88cff5ce4c1ca2ad19cc87e9841b030 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 20 Oct 2017 21:47:49 -0400
Subject: [PATCH] Firework APIs


diff --git a/src/main/java/net/minecraft/entity/item/EntityFireworkRocket.java b/src/main/java/net/minecraft/entity/item/EntityFireworkRocket.java
index 0fb5150..e0bfc16 100644
--- a/src/main/java/net/minecraft/entity/item/EntityFireworkRocket.java
+++ b/src/main/java/net/minecraft/entity/item/EntityFireworkRocket.java
@@ -23,6 +23,8 @@ import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.World;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 
+import java.util.UUID;
+
 public class EntityFireworkRocket extends Entity {
     public static final DataParameter<ItemStack> FIREWORK_ITEM =
             EntityDataManager.<ItemStack>createKey(EntityFireworkRocket.class, DataSerializers.ITEM_STACK);
@@ -31,6 +33,7 @@ public class EntityFireworkRocket extends Entity {
     private int fireworkAge;
     public int lifetime;
     public EntityLivingBase boostedEntity;
+    public UUID spawningEntity; // Paper
 
     public EntityFireworkRocket(World worldIn) {
         super(worldIn);
@@ -230,6 +233,12 @@ public class EntityFireworkRocket extends Entity {
         if (!itemstack.isEmpty()) {
             compound.setTag("FireworksItem", itemstack.writeToNBT(new NBTTagCompound()));
         }
+
+        // Paper start
+        if (spawningEntity != null) {
+            compound.setUniqueId("SpawningEntity", spawningEntity);
+        }
+        // Paper end
     }
 
     public void readEntityFromNBT(NBTTagCompound compound) {
@@ -244,6 +253,12 @@ public class EntityFireworkRocket extends Entity {
                 this.dataManager.set(FIREWORK_ITEM, itemstack);
             }
         }
+
+        // Paper start
+        if (compound.hasUniqueId("SpawningEntity")) {
+            spawningEntity = compound.getUniqueId("SpawningEntity");
+        }
+        // Paper end
     }
 
     public boolean canBeAttackedWithItem() {
diff --git a/src/main/java/net/minecraft/item/ItemFirework.java b/src/main/java/net/minecraft/item/ItemFirework.java
index 6c590ca..f530afa 100644
--- a/src/main/java/net/minecraft/item/ItemFirework.java
+++ b/src/main/java/net/minecraft/item/ItemFirework.java
@@ -21,6 +21,7 @@ public class ItemFirework extends Item {
                             (double) ((float) pos.getY() + hitY),
                             (double) ((float) pos.getZ() + hitZ),
                             var9);
+            var10.spawningEntity = player.getUniqueID(); // Paper
             worldIn.spawnEntity(var10);
 
             if (!player.capabilities.isCreativeMode) {
@@ -37,6 +38,7 @@ public class ItemFirework extends Item {
 
             if (!worldIn.isRemote) {
                 EntityFireworkRocket var5 = new EntityFireworkRocket(worldIn, var4, playerIn);
+                var5.spawningEntity = playerIn.getUniqueID(); // Paper
                 worldIn.spawnEntity(var5);
 
                 if (!playerIn.capabilities.isCreativeMode) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
index e2d6071..8a26331 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
@@ -1,6 +1,9 @@
 package org.bukkit.craftbukkit.entity;
 
 import java.util.Random;
+import java.util.UUID;
+
+import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.item.EntityFireworkRocket;
 import net.minecraft.init.Items;
 import net.minecraft.item.ItemStack;
@@ -9,6 +12,7 @@ import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.bukkit.entity.EntityType;
 import org.bukkit.entity.Firework;
+import org.bukkit.entity.LivingEntity;
 import org.bukkit.inventory.meta.FireworkMeta;
 
 public class CraftFirework extends CraftEntity implements Firework {
@@ -56,4 +60,18 @@ public class CraftFirework extends CraftEntity implements Firework {
     public void detonate() {
         this.getHandle().lifetime = 0;
     }
+
+    // Paper start
+
+    @Override
+    public UUID getSpawningEntity() {
+        return getHandle().spawningEntity;
+    }
+
+    @Override
+    public LivingEntity getBoostedEntity() {
+        EntityLivingBase boostedEntity = getHandle().boostedEntity;
+        return boostedEntity != null ? (LivingEntity) boostedEntity.getBukkitEntity() : null;
+    }
+    // Paper end
 }
-- 
2.18.0

