From fb6120bd2734edaf4eb4294d394b3f43013a2348 Mon Sep 17 00:00:00 2001
From: Alfie Cleveland <alfeh@me.com>
Date: Fri, 20 Oct 2017 22:06:53 -0400
Subject: [PATCH] Don't allow entities to ride themselves - #572


diff --git a/src/main/java/net/minecraft/entity/Entity.java b/src/main/java/net/minecraft/entity/Entity.java
index d6b805c..a9d1306 100644
--- a/src/main/java/net/minecraft/entity/Entity.java
+++ b/src/main/java/net/minecraft/entity/Entity.java
@@ -1929,6 +1929,7 @@ public abstract class Entity implements ICommandSender {
     }
 
     protected void addPassenger(Entity passenger) {
+        if (passenger == this) throw new IllegalArgumentException("Entities cannot become a passenger of themselves"); // Paper - issue 572
         if (passenger.getRidingEntity() != this) {
             throw new IllegalStateException("Use x.startRiding(y), not y.addPassenger(x)");
         } else {
-- 
2.18.0

