From 1c581c3f138d2ec2cefb93288e3ab91cede76667 Mon Sep 17 00:00:00 2001
From: Michael Himing <mhiming@gmail.com>
Date: Fri, 20 Oct 2017 22:08:10 -0400
Subject: [PATCH] Fix block break desync


diff --git a/src/main/java/net/minecraft/network/NetHandlerPlayServer.java b/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
index 9eead06..fb17aa9 100644
--- a/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
+++ b/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
@@ -1050,6 +1050,7 @@ public class NetHandlerPlayServer implements INetHandlerPlayServer, ITickable {
                     double d3 = d0 * d0 + d1 * d1 + d2 * d2;
 
                     if (d3 > 36.0D) {
+                        this.sendPacket(new SPacketBlockChange(worldserver, blockposition)); // Paper - Fix block break desync
                         return;
                     } else if (blockposition.getY() >= this.serverController.getBuildLimit()) {
                         return;
-- 
2.18.0

