From 8e950af5d488a389e2d40d52aa3c8882e7745f64 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Fri, 20 Oct 2017 22:10:29 -0400
Subject: [PATCH] Assign the world in MapGenStrongHold


diff --git a/src/main/java/net/minecraft/world/gen/structure/MapGenStronghold.java b/src/main/java/net/minecraft/world/gen/structure/MapGenStronghold.java
index f79ff3d..d376fe1 100644
--- a/src/main/java/net/minecraft/world/gen/structure/MapGenStronghold.java
+++ b/src/main/java/net/minecraft/world/gen/structure/MapGenStronghold.java
@@ -51,6 +51,7 @@ public class MapGenStronghold extends MapGenStructure {
     }
 
     public BlockPos getNearestStructurePos(World worldIn, BlockPos pos, boolean findUnexplored) {
+        this.world = worldIn; // Paper
         if (!this.ranBiomeCheck) {
             this.generatePositions();
             this.ranBiomeCheck = true;
-- 
2.18.0

