From c6f5cd6f08777cde02909d9d7203f0bae98e49a2 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sun, 29 Oct 2017 17:25:09 -0400
Subject: [PATCH] Add fromBottle flag to Experience Orbs


diff --git a/src/main/java/net/minecraft/entity/item/EntityExpBottle.java b/src/main/java/net/minecraft/entity/item/EntityExpBottle.java
index daba9d6..6a77fc3 100644
--- a/src/main/java/net/minecraft/entity/item/EntityExpBottle.java
+++ b/src/main/java/net/minecraft/entity/item/EntityExpBottle.java
@@ -45,7 +45,7 @@ public class EntityExpBottle extends EntityThrowable {
             while (i > 0) {
                 int j = EntityXPOrb.getXPSplit(i);
                 i -= j;
-                this.world.spawnEntity(new EntityXPOrb(this.world, this.posX, this.posY, this.posZ, j));
+                this.world.spawnEntity(new EntityXPOrb(this.world, this.posX, this.posY, this.posZ, j, true)); // Paper - add fromBottle flag
             }
 
             this.setDead();
diff --git a/src/main/java/net/minecraft/entity/item/EntityXPOrb.java b/src/main/java/net/minecraft/entity/item/EntityXPOrb.java
index b683adf..e3e77df 100644
--- a/src/main/java/net/minecraft/entity/item/EntityXPOrb.java
+++ b/src/main/java/net/minecraft/entity/item/EntityXPOrb.java
@@ -27,6 +27,14 @@ public class EntityXPOrb extends Entity {
     public int xpValue;
     private EntityPlayer closestPlayer;
     private int xpTargetColor;
+    // Paper start
+    private boolean fromBottle = false;
+
+    public EntityXPOrb(World worldIn, double x, double y, double z, int expValue, boolean fromBottle) {
+        this(worldIn, x, y, z, expValue);
+        this.fromBottle = fromBottle;
+    }
+    // Paper end
 
     public EntityXPOrb(World worldIn, double x, double y, double z, int expValue) {
         super(worldIn);
@@ -214,6 +222,12 @@ public class EntityXPOrb extends Entity {
         return this.xpValue;
     }
 
+    // Paper start
+    public boolean isFromBottle() {
+        return fromBottle;
+    }
+    // Paper end
+
     public static int getXPSplit(int expValue) {
         if (expValue > 162670129) {
             return expValue - 100000;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftExperienceOrb.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftExperienceOrb.java
index 19e13a2..b894848 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftExperienceOrb.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftExperienceOrb.java
@@ -29,4 +29,11 @@ public class CraftExperienceOrb extends CraftEntity implements ExperienceOrb {
     public EntityType getType() {
         return EntityType.EXPERIENCE_ORB;
     }
+
+    // Paper start
+    @Override
+    public boolean isFromBottle() {
+        return getHandle().isFromBottle();
+    }
+    // Paper end
 }
-- 
2.14.3.windows.1

