From 54ecb35b9636c3b127d0230c325207e685bea99d Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sun, 29 Oct 2017 18:07:12 -0400
Subject: [PATCH] Do not let armorstands drown


diff --git a/src/main/java/net/minecraft/entity/item/EntityArmorStand.java b/src/main/java/net/minecraft/entity/item/EntityArmorStand.java
index c7bbfd2..bd22601 100644
--- a/src/main/java/net/minecraft/entity/item/EntityArmorStand.java
+++ b/src/main/java/net/minecraft/entity/item/EntityArmorStand.java
@@ -135,6 +135,13 @@ public class EntityArmorStand extends EntityLivingBase {
         this.dataManager.register(RIGHT_LEG_ROTATION, DEFAULT_RIGHTLEG_ROTATION);
     }
 
+    // Paper start - Don't let ArmorStands drown - probably a micro-optimization for most servers
+    @Override
+    public boolean canBreatheUnderwater() {
+        return true;
+    }
+    // Paper end
+
     public Iterable<ItemStack> getHeldEquipment() {
         return this.handItems;
     }
-- 
2.14.3

