From f8f4ed9ff3874ea3759368254337e48a0d91ae4c Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sun, 29 Oct 2017 18:07:12 -0400
Subject: [PATCH] Do not let armorstands drown


diff --git a/src/main/java/net/minecraft/entity/item/EntityArmorStand.java b/src/main/java/net/minecraft/entity/item/EntityArmorStand.java
index 1ab5671..606f681 100644
--- a/src/main/java/net/minecraft/entity/item/EntityArmorStand.java
+++ b/src/main/java/net/minecraft/entity/item/EntityArmorStand.java
@@ -135,6 +135,13 @@ public class EntityArmorStand extends EntityLivingBase {
         this.dataManager.register(RIGHT_LEG_ROTATION, DEFAULT_RIGHTLEG_ROTATION);
     }
 
+    // Paper start - Don't let ArmorStands drown - probably a micro-optimization for most servers
+    @Override
+    public boolean canBreatheUnderwater() {
+        return true;
+    }
+    // Paper end
+
     public Iterable<ItemStack> getHeldEquipment() {
         return this.handItems;
     }
-- 
2.18.0

