From 9507dcf779ee04923a560cc815374d6757182c94 Mon Sep 17 00:00:00 2001
From: kashike <kashike@vq.lc>
Date: Sun, 29 Oct 2017 18:09:57 -0400
Subject: [PATCH] Fix NFE when attempting to read EMPTY ItemStack

Thanks @gabizou

diff --git a/src/main/java/net/minecraft/item/ItemStack.java b/src/main/java/net/minecraft/item/ItemStack.java
index 478febf..8361ff5 100644
--- a/src/main/java/net/minecraft/item/ItemStack.java
+++ b/src/main/java/net/minecraft/item/ItemStack.java
@@ -147,7 +147,7 @@ public final class ItemStack {
     }
 
     public void load(NBTTagCompound nbttagcompound) {
-        this.item = Item.getByNameOrId(nbttagcompound.getString("id"));
+        this.item = nbttagcompound.hasKey("id", 8) ? Item.getByNameOrId(nbttagcompound.getString("id")) : Item.getItemFromBlock(Blocks.AIR); // Paper - fix NumberFormatException caused by attempting to read an EMPTY ItemStack
         this.stackSize = nbttagcompound.getByte("Count");
         this.setItemDamage(nbttagcompound.getShort("Damage"));
 
-- 
2.14.3.windows.1

