From 8f60def39e34a4525e5d088f50d0644f2da14388 Mon Sep 17 00:00:00 2001
From: kashike <kashike@vq.lc>
Date: Sun, 29 Oct 2017 18:09:57 -0400
Subject: [PATCH] Fix NFE when attempting to read EMPTY ItemStack

Thanks @gabizou

diff --git a/src/main/java/net/minecraft/item/ItemStack.java b/src/main/java/net/minecraft/item/ItemStack.java
index d75f8da..0db4cac 100644
--- a/src/main/java/net/minecraft/item/ItemStack.java
+++ b/src/main/java/net/minecraft/item/ItemStack.java
@@ -163,7 +163,7 @@ public final class ItemStack {
     }
 
     public void load(NBTTagCompound nbttagcompound) {
-        this.item = Item.getByNameOrId(nbttagcompound.getString("id"));
+        this.item = nbttagcompound.hasKey("id", 8) ? Item.getByNameOrId(nbttagcompound.getString("id")) : Item.getItemFromBlock(Blocks.AIR); // Paper - fix NumberFormatException caused by attempting to read an EMPTY ItemStack
         this.stackSize = nbttagcompound.getByte("Count");
         this.setItemDamage(nbttagcompound.getShort("Damage"));
 
-- 
2.18.0

