From f84d92c02b113509dd23b0530d58cafbaee2e909 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sun, 29 Oct 2017 20:15:50 -0400
Subject: [PATCH] PlayerAttemptPickupItemEvent


diff --git a/src/main/java/net/minecraft/entity/item/EntityItem.java b/src/main/java/net/minecraft/entity/item/EntityItem.java
index b1cafee..d0513f9 100644
--- a/src/main/java/net/minecraft/entity/item/EntityItem.java
+++ b/src/main/java/net/minecraft/entity/item/EntityItem.java
@@ -31,6 +31,7 @@ import org.apache.logging.log4j.Logger;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.entity.Player;
 import org.bukkit.event.entity.EntityPickupItemEvent;
+import org.bukkit.event.player.PlayerAttemptPickupItemEvent;
 import org.bukkit.event.player.PlayerPickupItemEvent;
 
 // Paper start - implement HopperPusher
@@ -359,6 +360,13 @@ public class EntityItem extends Entity implements HopperPusher {
             ItemStack itemstack = this.getItem();
             Item item = itemstack.getItem();
             int i = itemstack.getCount();
+
+            // Paper start
+            if (this.pickupDelay <= 0) {
+                this.world.getServer().getPluginManager().callEvent(new PlayerAttemptPickupItemEvent((org.bukkit.entity.Player) entityIn.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity()));
+            }
+            // Paper end
+
             int canHold = entityIn.inventory.canHold(itemstack);
             int remaining = i - canHold;
             boolean flyAtPlayer = false; // Paper
-- 
2.14.3.windows.1

