From bc58a7704213d0d5bbed43fd61bb3b5ca3b330a4 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 29 Oct 2017 20:28:11 -0400
Subject: [PATCH] Shoulder Entities Release API


diff --git a/src/main/java/net/minecraft/entity/player/EntityPlayer.java b/src/main/java/net/minecraft/entity/player/EntityPlayer.java
index 7737dae..caa3f72 100644
--- a/src/main/java/net/minecraft/entity/player/EntityPlayer.java
+++ b/src/main/java/net/minecraft/entity/player/EntityPlayer.java
@@ -1892,29 +1892,61 @@ public abstract class EntityPlayer extends EntityLivingBase {
     }
 
     public void spawnShoulderEntities() {
-        if (this.spawnEntityFromShoulder(this.getLeftShoulderEntity())) {
+        if (this.spawnShoulderEntity(this.getLeftShoulderEntity())) {
             this.setLeftShoulderEntity(new NBTTagCompound());
         }
 
-        if (this.spawnEntityFromShoulder(this.getRightShoulderEntity())) {
+        if (this.spawnShoulderEntity(this.getRightShoulderEntity())) {
             this.setRightShoulderEntity(new NBTTagCompound());
         }
     }
 
-    private boolean spawnEntityFromShoulder(@Nullable NBTTagCompound nbttagcompound) {
-        if (!this.world.isRemote && !nbttagcompound.hasNoTags()) {
+    // Paper start
+    @Nullable
+    public Entity releaseLeftShoulderEntity() {
+        Entity entity = this.spawnEntityFromShoulder0(this.getLeftShoulderEntity());
+        if (entity != null) {
+            this.setLeftShoulderEntity(new NBTTagCompound());
+        }
+        return entity;
+    }
+
+    @Nullable
+    public Entity releaseRightShoulderEntity() {
+        Entity entity = this.spawnEntityFromShoulder0(this.getRightShoulderEntity());
+        if (entity != null) {
+            this.setRightShoulderEntity(new NBTTagCompound());
+        }
+        return entity;
+    }
+
+    // Paper - incase any plugins used NMS to call this... old method signature to avoid other diff
+    private boolean spawnShoulderEntity(@Nullable NBTTagCompound nbttagcompound) {
+        return spawnEntityFromShoulder0(nbttagcompound) != null;
+    }
+    // Paper - Moved to new method that now returns entity, and properly null checks
+    @Nullable
+    private Entity spawnEntityFromShoulder0(@Nullable NBTTagCompound nbttagcompound) { // Paper - return Entity
+        if (!this.world.isRemote &&  nbttagcompound != null && !nbttagcompound.hasNoTags()) {
             Entity entity = EntityList.createEntityFromNBT(nbttagcompound, this.world);
 
+            if (entity == null) { // Paper - null check
+                return null;
+            }
+
             if (entity instanceof EntityTameable) {
                 ((EntityTameable) entity).setOwnerId(this.entityUniqueID);
             }
 
             entity.setPosition(this.posX, this.posY + 0.699999988079071D, this.posZ);
-            return this.world.addEntity(entity, SpawnReason.SHOULDER_ENTITY);
-        } else {
-            return true;
+            if (this.world.addEntity(entity, SpawnReason.SHOULDER_ENTITY)) {
+                return entity;
+            }
         }
+
+        return null;
     }
+    // Paper end
 
     public abstract boolean isSpectator();
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index a1619dc..3dc10a1 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -482,6 +482,32 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         this.getHandle().getCooldownTracker().setCooldown(CraftMagicNumbers.getItem(material), ticks);
     }
 
+    // Paper start
+    @Override
+    public org.bukkit.entity.Entity releaseLeftShoulderEntity() {
+        if (!getHandle().getLeftShoulderEntity().hasNoTags()) {
+            net.minecraft.entity.Entity entity = getHandle().releaseLeftShoulderEntity();
+            if (entity != null) {
+                return entity.getBukkitEntity();
+            }
+        }
+
+        return null;
+    }
+
+    @Override
+    public org.bukkit.entity.Entity releaseRightShoulderEntity() {
+        if (!getHandle().getRightShoulderEntity().hasNoTags()) {
+            net.minecraft.entity.Entity entity = getHandle().releaseRightShoulderEntity();
+            if (entity != null) {
+                return entity.getBukkitEntity();
+            }
+        }
+
+        return null;
+    }
+    // Paper end
+
     public Entity getShoulderEntityLeft() {
         if (!this.getHandle().getLeftShoulderEntity().hasNoTags()) {
             net.minecraft.entity.Entity shoulder =
-- 
2.14.3

