From 227a055cbd6ce2122567f4b825060e9161a28b77 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sun, 29 Oct 2017 20:50:07 -0400
Subject: [PATCH] Block player logins during server shutdown


diff --git a/src/main/java/net/minecraft/server/network/NetHandlerLoginServer.java b/src/main/java/net/minecraft/server/network/NetHandlerLoginServer.java
index 8ba7ea5..66b2315 100644
--- a/src/main/java/net/minecraft/server/network/NetHandlerLoginServer.java
+++ b/src/main/java/net/minecraft/server/network/NetHandlerLoginServer.java
@@ -64,6 +64,12 @@ public class NetHandlerLoginServer implements INetHandlerLoginServer, ITickable
     }
 
     public void update() {
+        // Paper start - Do not allow logins while the server is shutting down
+        if (!MinecraftServer.getServer().isServerRunning()) {
+            this.disconnect(new TextComponentString(org.spigotmc.SpigotConfig.restartMessage));
+            return;
+        }
+        // Paper end
         if (this.currentLoginState == NetHandlerLoginServer.LoginState.READY_TO_ACCEPT) {
             this.tryAcceptPlayer();
         } else if (this.currentLoginState == NetHandlerLoginServer.LoginState.DELAY_ACCEPT) {
-- 
2.14.3.windows.1

