From d570ae83f79f8059ee7eb25405abcff6c576fa67 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sun, 29 Oct 2017 21:30:07 -0400
Subject: [PATCH] Ocelot despawns should honor nametags and leash


diff --git a/src/main/java/net/minecraft/entity/passive/EntityOcelot.java b/src/main/java/net/minecraft/entity/passive/EntityOcelot.java
index 47eb7de..486f8e7 100644
--- a/src/main/java/net/minecraft/entity/passive/EntityOcelot.java
+++ b/src/main/java/net/minecraft/entity/passive/EntityOcelot.java
@@ -97,7 +97,7 @@ public class EntityOcelot extends EntityTameable {
     }
 
     public boolean canDespawn() {
-        return !this.isTamed();
+        return !this.isTamed() && !this.hasCustomName() && !this.getLeashed(); // Paper - Honor name and leash status
     }
 
     protected void applyEntityAttributes() {
-- 
2.18.0

