From 9568d87075cf84325f7539fbc6708559028f0761 Mon Sep 17 00:00:00 2001
From: mezz <tehgeek@gmail.com>
Date: Sun, 29 Oct 2017 21:36:28 -0400
Subject: [PATCH] Fix MC-117075: TE Unload Lag Spike


diff --git a/src/main/java/net/minecraft/world/World.java b/src/main/java/net/minecraft/world/World.java
index 9802157..7f535a3 100644
--- a/src/main/java/net/minecraft/world/World.java
+++ b/src/main/java/net/minecraft/world/World.java
@@ -1561,7 +1561,11 @@ public abstract class World implements IBlockAccess {
         this.timings.tileEntityTick.startTiming();
 
         if (!this.tileEntitiesToBeRemoved.isEmpty()) {
-            this.tickableTileEntities.removeAll(this.tileEntitiesToBeRemoved);
+            // Paper start - Use alternate implementation with faster contains
+            java.util.Set<TileEntity> toRemove = java.util.Collections.newSetFromMap(new java.util.IdentityHashMap<>());
+            toRemove.addAll(tileEntitiesToBeRemoved);
+            this.tickableTileEntities.removeAll(toRemove);
+            // Paper end
             //this.loadedTileEntityList.removeAll(this.tileEntitiesToBeRemoved); // Paper - Remove unused list
             this.tileEntitiesToBeRemoved.clear();
         }
-- 
2.18.0

