From eba8d03d539e85ad3608a789be72d1cfdf7dd810 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sun, 29 Oct 2017 21:41:04 -0400
Subject: [PATCH] LivingEntity#setKiller


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 1597005..928d4da 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -254,6 +254,16 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         return this.getHandle().attackingPlayer == null ? null : (Player) this.getHandle().attackingPlayer.getBukkitEntity();
     }
 
+    // Paper start
+    @Override
+    public void setKiller(Player killer) {
+        EntityPlayer entityPlayer = killer == null ? null : ((CraftPlayer) killer).getHandle();
+        getHandle().attackingPlayer = entityPlayer;
+        getHandle().revengeTarget = entityPlayer;
+        getHandle().recentlyHit = entityPlayer == null ? 0 : 100; // 100 value taken from EntityLiving#damageEntity
+    }
+    // Paper end
+
     public boolean addPotionEffect(PotionEffect effect) {
         return this.addPotionEffect(effect, false);
     }
-- 
2.14.3.windows.1

