From d8c2340d667282f1c53a6d1b5b88b9affd650d3d Mon Sep 17 00:00:00 2001
From: Brokkonaut <hannos17@gmx.de>
Date: Tue, 31 Oct 2017 03:26:18 +0100
Subject: [PATCH] Send attack SoundEffects only to players who can see the
 attacker


diff --git a/src/main/java/net/minecraft/entity/player/EntityPlayer.java b/src/main/java/net/minecraft/entity/player/EntityPlayer.java
index caa3f72..4b13cc0 100644
--- a/src/main/java/net/minecraft/entity/player/EntityPlayer.java
+++ b/src/main/java/net/minecraft/entity/player/EntityPlayer.java
@@ -52,6 +52,7 @@ import net.minecraft.network.datasync.DataParameter;
 import net.minecraft.network.datasync.DataSerializers;
 import net.minecraft.network.datasync.EntityDataManager;
 import net.minecraft.network.play.server.SPacketEntityVelocity;
+import net.minecraft.network.play.server.SPacketSoundEffect;
 import net.minecraft.scoreboard.ScorePlayerTeam;
 import net.minecraft.scoreboard.Scoreboard;
 import net.minecraft.stats.StatBase;
@@ -1018,6 +1019,15 @@ public abstract class EntityPlayer extends EntityLivingBase {
         this.rideCooldown = 0;
     }
 
+    // Paper start - send SoundEffect to everyone who can see fromEntity
+    private static void sendSoundEffect(EntityPlayer fromEntity, double x, double y, double z, SoundEvent soundEffect, SoundCategory soundCategory, float volume, float pitch) {
+        fromEntity.world.playSound(fromEntity, x, y, z, soundEffect, soundCategory, volume, pitch); // This will not send the effect to the entity himself
+        if (fromEntity instanceof EntityPlayerMP) {
+            ((EntityPlayerMP) fromEntity).connection.sendPacket(new SPacketSoundEffect(soundEffect, soundCategory, x, y, z, volume, pitch));
+        }
+    }
+    // Paper end
+
     public void attackTargetEntityWithCurrentItem(Entity targetEntity) {
         if (targetEntity.canBeAttackedWithItem() && !targetEntity.hitByEntity(this)) {
             float f = (float) this.getEntityAttribute(SharedMonsterAttributes.ATTACK_DAMAGE).getAttributeValue();
@@ -1043,8 +1053,10 @@ public abstract class EntityPlayer extends EntityLivingBase {
                 int i = b0 + EnchantmentHelper.getKnockbackModifier(this);
 
                 if (this.isSprinting() && flag) {
-                    this.world.playSound(
-                            (EntityPlayer) null,
+                    // Paper start - send while respecting visibility
+                    sendSoundEffect(
+                            this,
+                            // Paper end
                             this.posX,
                             this.posY,
                             this.posZ,
@@ -1147,8 +1159,10 @@ public abstract class EntityPlayer extends EntityLivingBase {
                             }
                         }
 
-                        this.world.playSound(
-                                (EntityPlayer) null,
+                        // Paper start - send while respecting visibility
+                        sendSoundEffect(
+                                this,
+                                // Paper end
                                 this.posX,
                                 this.posY,
                                 this.posZ,
@@ -1182,8 +1196,10 @@ public abstract class EntityPlayer extends EntityLivingBase {
                     }
 
                     if (flag2) {
-                        this.world.playSound(
-                                (EntityPlayer) null,
+                        // Paper start - send while respecting visibility
+                        sendSoundEffect(
+                                this,
+                                // Paper end
                                 this.posX,
                                 this.posY,
                                 this.posZ,
@@ -1196,8 +1212,10 @@ public abstract class EntityPlayer extends EntityLivingBase {
 
                     if (!flag2 && !flag3) {
                         if (flag) {
-                            this.world.playSound(
-                                    (EntityPlayer) null,
+                            // Paper start - send while respecting visibility
+                            sendSoundEffect(
+                                    this,
+                                    // Paper end
                                     this.posX,
                                     this.posY,
                                     this.posZ,
@@ -1206,8 +1224,10 @@ public abstract class EntityPlayer extends EntityLivingBase {
                                     1.0F,
                                     1.0F);
                         } else {
-                            this.world.playSound(
-                                    (EntityPlayer) null,
+                            // Paper start - send while respecting visibility
+                            sendSoundEffect(
+                                    this,
+                                    // Paper end
                                     this.posX,
                                     this.posY,
                                     this.posZ,
@@ -1280,8 +1300,10 @@ public abstract class EntityPlayer extends EntityLivingBase {
 
                     this.addExhaustion(this.world.spigotConfig.combatExhaustion);
                 } else {
-                    this.world.playSound(
-                            (EntityPlayer) null,
+                    // Paper start - send while respecting visibility
+                    sendSoundEffect(
+                            this,
+                            // Paper end
                             this.posX,
                             this.posY,
                             this.posZ,
-- 
2.14.3

